// react-pretext-editor/dist/index.js
var nn = Object.defineProperty;
var rn = (n, e, t) => e in n ? nn(n, e, { enumerable: !0, configurable: !0, writable: !0, value: t }) : n[e] = t;
var Y = (n, e, t) => rn(n, typeof e != "symbol" ? e + "" : e, t);
import { jsxs as K, jsx as v, Fragment as sn } from "react/jsx-runtime";
import Ke, { useContext as on, createContext as ln, forwardRef as st, createElement as Je, useState as ke, useRef as ue, useEffect as Me, useCallback as ae, useImperativeHandle as an } from "react";
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const Pt = (...n) => n.filter((e, t, r) => !!e && e.trim() !== "" && r.indexOf(e) === t).join(" ").trim();
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const cn = (n) => n.replace(/([a-z0-9])([A-Z])/g, "$1-$2").toLowerCase();
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const pn = (n) => n.replace(
  /^([A-Z])|[\s-_]+(\w)/g,
  (e, t, r) => r ? r.toUpperCase() : t.toLowerCase()
);
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const Ct = (n) => {
  const e = pn(n);
  return e.charAt(0).toUpperCase() + e.slice(1);
};
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
var Ve = {
  xmlns: "http://www.w3.org/2000/svg",
  width: 24,
  height: 24,
  viewBox: "0 0 24 24",
  fill: "none",
  stroke: "currentColor",
  strokeWidth: 2,
  strokeLinecap: "round",
  strokeLinejoin: "round"
};
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const hn = (n) => {
  for (const e in n)
    if (e.startsWith("aria-") || e === "role" || e === "title")
      return !0;
  return !1;
}, un = ln({}), dn = () => on(un), gn = st(
  ({ color: n, size: e, strokeWidth: t, absoluteStrokeWidth: r, className: s = "", children: o, iconNode: l, ...i }, c) => {
    const {
      size: p = 24,
      strokeWidth: f = 2,
      absoluteStrokeWidth: y = !1,
      color: b = "currentColor",
      className: d = ""
    } = dn() ?? {}, I = r ?? y ? Number(t ?? f) * 24 / Number(e ?? p) : t ?? f;
    return Je(
      "svg",
      {
        ref: c,
        ...Ve,
        width: e ?? p ?? Ve.width,
        height: e ?? p ?? Ve.height,
        stroke: n ?? b,
        strokeWidth: I,
        className: Pt("lucide", d, s),
        ...!o && !hn(i) && { "aria-hidden": "true" },
        ...i
      },
      [
        ...l.map(([z, T]) => Je(z, T)),
        ...Array.isArray(o) ? o : [o]
      ]
    );
  }
);
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const $e = (n, e) => {
  const t = st(
    ({ className: r, ...s }, o) => Je(gn, {
      ref: o,
      iconNode: e,
      className: Pt(
        `lucide-${cn(Ct(n))}`,
        `lucide-${n}`,
        r
      ),
      ...s
    })
  );
  return t.displayName = Ct(n), t;
};
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const fn = [
  ["path", { d: "m15 16 2.536-7.328a1.02 1.02 1 0 1 1.928 0L22 16", key: "xik6mr" }],
  ["path", { d: "M15.697 14h5.606", key: "1stdlc" }],
  ["path", { d: "m2 16 4.039-9.69a.5.5 0 0 1 .923 0L11 16", key: "d5nyq2" }],
  ["path", { d: "M3.304 13h6.392", key: "1q3zxz" }]
], xn = $e("a-large-small", fn);
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const kn = [
  ["rect", { width: "18", height: "18", x: "3", y: "3", rx: "2", key: "afitv7" }],
  ["path", { d: "M12 3v18", key: "108xh3" }]
], mn = $e("columns-2", kn);
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const bn = [
  [
    "path",
    {
      d: "M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0",
      key: "1nclc0"
    }
  ],
  ["circle", { cx: "12", cy: "12", r: "3", key: "1v7zrd" }]
], yn = $e("eye", bn);
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const wn = [
  ["rect", { width: "7", height: "7", x: "3", y: "3", rx: "1", key: "1g98yp" }],
  ["rect", { width: "7", height: "7", x: "14", y: "3", rx: "1", key: "6d4xhi" }],
  ["rect", { width: "7", height: "7", x: "14", y: "14", rx: "1", key: "nxv5o0" }],
  ["rect", { width: "7", height: "7", x: "3", y: "14", rx: "1", key: "1bb6yr" }]
], Sn = $e("layout-grid", wn);
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const Cn = [
  [
    "path",
    {
      d: "M21.174 6.812a1 1 0 0 0-3.986-3.987L3.842 16.174a2 2 0 0 0-.5.83l-1.321 4.352a.5.5 0 0 0 .623.622l4.353-1.32a2 2 0 0 0 .83-.497z",
      key: "1a8usu"
    }
  ]
], Rn = $e("pen", Cn);
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const $n = [
  ["path", { d: "M5 12h14", key: "1ays0h" }],
  ["path", { d: "M12 5v14", key: "s699le" }]
], zn = $e("plus", $n);
/**
 * @license lucide-react v1.7.0 - ISC
 *
 * This source code is licensed under the ISC license.
 * See the LICENSE file in the root directory of this source tree.
 */
const vn = [
  ["path", { d: "M18 6 6 18", key: "1bl5f8" }],
  ["path", { d: "m6 6 12 12", key: "d8bk6v" }]
], In = $e("x", vn);
let Te = null, Be = null;
function Ln() {
  return Te ? Promise.resolve(Te) : Be || (Be = import("./layout-BDxeR5zj.js").then((n) => (Te = n, Te)), Be);
}
function Tn() {
  const [n, e] = ke(!!Te), t = ue(Te);
  return Me(() => {
    t.current || Ln().then((r) => {
      t.current = r, e(!0);
    });
  }, []), n ? t.current : null;
}
const An = {
  bodyFont: "16px Lato, sans-serif",
  headingFont: "bold 24px Lato, sans-serif",
  h1Font: "bold 32px Lato, sans-serif",
  h2Font: "bold 24px Lato, sans-serif",
  h3Font: "bold 20px Lato, sans-serif",
  dropCapFont: "96px FloralCapitals, serif",
  bodyLineHeight: 26,
  headingLineHeight: 36,
  h1LineHeight: 42,
  h2LineHeight: 34,
  h3LineHeight: 28,
  blockGap: 16,
  imgPadding: 10,
  dropCap: !1,
  dropCapOffsetX: 0,
  dropCapOffsetY: 0,
  columns: 1,
  minColumnWidth: 300,
  defaultMode: "layout"
};
function Mn(n, e, t, r, s = 0) {
  const o = n.width * (n.aspectRatio || 1.2), l = n.y, i = n.y + o + t;
  if (s > 0) {
    if (e + s <= l || e >= i) return null;
  } else if (e < l || e >= i) return null;
  const c = n.x, p = { left: c - t, right: c + n.width + t };
  if (n.polygon && n.polygon.length >= 3) {
    const f = s > 0 ? [e, e + s * 0.5, e + s - 1] : [e];
    let y = 1 / 0, b = -1 / 0, d = !1;
    const I = n.polygon;
    for (const z of f) {
      const T = (z - n.y) / o;
      if (!(T < 0 || T > 1)) {
        d = !0;
        for (let J = 0; J < I.length; J++) {
          const ne = I[J], B = I[(J + 1) % I.length];
          if (ne.y <= T && B.y > T || B.y <= T && ne.y > T) {
            const E = (T - ne.y) / (B.y - ne.y), $ = ne.x + E * (B.x - ne.x), D = c + $ * n.width;
            y = Math.min(y, D), b = Math.max(b, D);
          }
        }
      }
    }
    return d ? y === 1 / 0 ? null : { left: y - t * 0.4, right: b + t } : p;
  }
  return p;
}
function Pn(n, e, t, r, s, o = 0) {
  const l = [];
  for (let p = 0; p < n.length; p++) {
    const f = Mn(n[p], e, r, s == null ? void 0 : s[p], o);
    f && l.push(f);
  }
  if (l.length === 0) return [{ left: 0, right: t }];
  l.sort((p, f) => p.left - f.left);
  const i = [];
  let c = 0;
  for (const p of l)
    p.left > c && i.push({ left: c, right: Math.min(p.left, t) }), c = Math.max(c, p.right);
  return c < t && i.push({ left: c, right: t }), i.filter((p) => p.right - p.left > 80);
}
function En(n, e) {
  const t = e < 1 ? Math.sqrt(e) : e;
  return n.map((r, s) => ({
    index: s,
    x: r.x * e,
    y: r.y * e,
    width: r.width * t,
    aspectRatio: r.aspectRatio,
    polygon: r.polygon,
    url: r.url,
    alt: r.alt,
    filename: r.filename
  }));
}
function Fn(n, e, t, r) {
  let s = 0;
  for (const o of n) {
    if (o.type === "hr") {
      s += 20 + r.blockGap;
      continue;
    }
    if (o.type === "code") {
      s += o.text.split(`
`).length * 20 + 20 + r.blockGap;
      continue;
    }
    if (o.type === "list") {
      for (const y of o.items || []) {
        const b = t.prepareWithSegments(y.text, r.bodyFont);
        let d = { segmentIndex: 0, graphemeIndex: 0 };
        for (; ; ) {
          const I = t.layoutNextLine(b, d, e - 24);
          if (!I) break;
          s += r.bodyLineHeight, d = I.end;
        }
        s += 4;
      }
      s += r.blockGap;
      continue;
    }
    const l = o.type === "heading", i = l ? o.tag === "h1" ? r.h1Font : o.tag === "h3" ? r.h3Font : r.h2Font : r.bodyFont, c = l ? o.tag === "h1" ? r.h1LineHeight : o.tag === "h3" ? r.h3LineHeight : r.h2LineHeight : r.bodyLineHeight;
    l && (s += r.blockGap);
    const p = t.prepareWithSegments(o.text, i);
    let f = { segmentIndex: 0, graphemeIndex: 0 };
    for (; ; ) {
      const y = t.layoutNextLine(p, f, e);
      if (!y) break;
      s += c, f = y.end;
    }
    s += r.blockGap;
  }
  return s;
}
function Hn(n, e, t, r, s = {}, o) {
  const l = { ...An, ...s }, i = 20;
  let c = l.columns;
  c > 1 && t / c < l.minColumnWidth && (c = Math.max(1, Math.floor(t / l.minColumnWidth)));
  const p = c > 1 ? (t - (c - 1) * i) / c : t, f = Array.from({ length: c }, () => []);
  for (const E of e)
    for (let $ = 0; $ < c; $++) {
      const D = $ * (p + i), h = D + p;
      E.x + E.width > D && E.x < h && f[$].push({ ...E, x: E.x - D });
    }
  const y = 24, b = 20, d = 3, I = (E, $) => {
    const D = [], h = (W, A, O, m, S, P, g, X, C, V, M = 1 / 0) => {
      const j = r.prepareWithSegments(W, A);
      let ie = { segmentIndex: 0, graphemeIndex: 0 }, ee = g, te = 0;
      for (; ; ) {
        if (c > 1 && ee >= E && X < c - 1)
          return { y: ee, cursor: ie, done: !1, overflow: !0, charOffset: te };
        const Z = ee < M ? m : 0, ce = Pn(S, ee, p, l.imgPadding, void 0, O);
        if (ce.length === 0) {
          ee += O;
          continue;
        }
        const me = ce.filter((fe) => Math.min(fe.right, p) - Math.max(fe.left, Z) >= 80);
        if (me.length === 0) {
          ee += O;
          continue;
        }
        let we = !1;
        for (const fe of me) {
          const be = Math.min(fe.right, p) - Math.max(fe.left, Z), Ge = Math.max(fe.left, Z), ve = r.layoutNextLine(j, ie, be);
          if (!ve)
            return we && (ee += O), { y: ee, cursor: ie, done: !0, overflow: !1, charOffset: te };
          D.push({
            type: "text",
            text: ve.text,
            x: Ge + P,
            y: ee,
            font: A,
            blockIndex: C,
            segments: V,
            charOffset: te
          }), te += ve.text.length, ie = ve.end, we = !0;
        }
        we && (ee += O);
      }
    };
    let w = 0, k = null, H = 0, Q = !0;
    e:
      for (let W = 0; W < c; W++) {
        const A = W * (p + i), O = f[W];
        let m = 0;
        for (; w < n.length; ) {
          const S = n[w];
          if (c > 1 && m >= E && W < c - 1) {
            H = Math.max(H, m);
            continue e;
          }
          switch (S.type) {
            case "heading": {
              k === null && (m += l.blockGap);
              const P = S.tag === "h1" ? l.h1Font : S.tag === "h3" ? l.h3Font : l.h2Font, g = S.tag === "h1" ? l.h1LineHeight : S.tag === "h3" ? l.h3LineHeight : l.h2LineHeight, X = h(
                S.text,
                P,
                g,
                0,
                O,
                A,
                m,
                W,
                w,
                S.segments
              );
              if (m = X.y, X.overflow) {
                k = X.cursor, H = Math.max(H, m);
                continue e;
              }
              break;
            }
            case "paragraph": {
              if (l.dropCap && Q && S.text.length > 0 && o) {
                Q = !1;
                const P = S.text[0], g = o(P, l.dropCapFont), X = g.width + 8, C = g.height * 0.78, V = l.dropCapOffsetX, M = l.dropCapOffsetY;
                D.push({
                  type: "dropCap",
                  char: P,
                  x: A + V,
                  y: m + M,
                  font: l.dropCapFont,
                  blockIndex: w
                });
                const j = S.text.slice(1), ie = S.segments ? (() => {
                  const Z = [...S.segments];
                  return Z.length > 0 && Z[0].text.length > 0 && (Z[0] = { ...Z[0], text: Z[0].text.slice(1) }), Z;
                })() : void 0, ee = m + C, te = h(
                  j,
                  l.bodyFont,
                  l.bodyLineHeight,
                  X,
                  O,
                  A,
                  m,
                  W,
                  w,
                  ie,
                  ee
                );
                if (m = te.y, te.overflow) {
                  k = te.cursor, H = Math.max(H, m);
                  continue e;
                }
              } else {
                S.type === "paragraph" && (Q = !1);
                const P = h(
                  S.text,
                  l.bodyFont,
                  l.bodyLineHeight,
                  0,
                  O,
                  A,
                  m,
                  W,
                  w,
                  S.segments
                );
                if (m = P.y, P.overflow) {
                  k = P.cursor, H = Math.max(H, m);
                  continue e;
                }
              }
              break;
            }
            case "list": {
              const P = S.items || [];
              for (let g = 0; g < P.length; g++) {
                const X = P[g], C = S.tag === "ol" ? `${g + 1}.` : "•";
                D.push({
                  type: "listBullet",
                  text: C,
                  x: A + 4,
                  y: m,
                  font: l.bodyFont,
                  blockIndex: w
                }), m = h(
                  X.text,
                  l.bodyFont,
                  l.bodyLineHeight,
                  y,
                  O,
                  A,
                  m,
                  W,
                  w,
                  X.segments
                ).y + 4;
              }
              break;
            }
            case "blockquote": {
              const P = m;
              D.push({
                type: "blockquoteBorder",
                x: A + 4,
                y: P,
                width: d,
                height: 0,
                // placeholder, updated below
                blockIndex: w
              });
              const g = h(
                S.text,
                l.bodyFont,
                l.bodyLineHeight,
                b,
                O,
                A,
                m,
                W,
                w,
                S.segments
              );
              if (m = g.y, D.length > 0) {
                D[D.length - 1 - (g.charOffset > 0 ? Math.ceil(g.charOffset / 40) : 0)];
                for (let X = D.length - 1; X >= 0; X--)
                  if (D[X].type === "blockquoteBorder" && D[X].blockIndex === w) {
                    D[X].height = m - P;
                    break;
                  }
              }
              break;
            }
            case "code": {
              const P = "14px 'SF Mono', 'Fira Code', Consolas, monospace", X = S.text.split(`
`), C = m;
              D.push({
                type: "codeBlock",
                x: A,
                y: C,
                width: p,
                height: X.length * 20 + 20,
                language: S.language,
                blockIndex: w
              }), m += 10;
              for (const V of X)
                D.push({
                  type: "text",
                  text: V,
                  x: A + 12,
                  y: m,
                  font: P,
                  blockIndex: w
                }), m += 20;
              m += 10;
              break;
            }
            case "hr": {
              m += 10, D.push({
                type: "hr",
                x: A,
                y: m,
                width: p,
                blockIndex: w
              }), m += 10;
              break;
            }
            default: {
              m = h(
                S.text,
                l.bodyFont,
                l.bodyLineHeight,
                0,
                O,
                A,
                m,
                W,
                w,
                S.segments
              ).y;
              break;
            }
          }
          w++, k = null, m += l.blockGap;
        }
        H = Math.max(H, m);
        break;
      }
    return { elements: D, maxColumnY: H };
  }, z = Fn(n, p, r, l);
  let T;
  if (c > 1) {
    let E = 0;
    for (const $ of e) {
      const D = $.width * ($.aspectRatio || 1.2), h = Math.min(1, $.width / p);
      E += D * h;
    }
    T = Math.ceil((z + E) / c);
  } else
    T = z;
  const { elements: J, maxColumnY: ne } = I(T);
  let B = ne;
  for (const E of e) {
    const $ = E.width * (E.aspectRatio || 1.2);
    B = Math.max(B, E.y + $), J.push({
      type: "image",
      x: E.x,
      y: E.y,
      width: E.width,
      imageIndex: E.index,
      url: E.url,
      alt: E.alt,
      polygon: E.polygon
    });
  }
  return { elements: J, totalHeight: B };
}
function _n(n, e) {
  if (n.breakpoints && n.breakpoints.length > 0) {
    const t = [...n.breakpoints].map((r, s) => ({ bp: r, idx: s })).sort((r, s) => r.bp.maxWidth - s.bp.maxWidth);
    for (const { bp: r, idx: s } of t)
      if (e <= r.maxWidth)
        return {
          images: r.images,
          columns: r.columns,
          editorWidth: r.editorWidth,
          fontFamily: r.fontFamily ?? n.fontFamily,
          fontSize: r.fontSize ?? n.fontSize,
          initialCap: r.initialCap ?? n.initialCap,
          initialCapFont: r.initialCapFont ?? n.initialCapFont,
          initialCapSize: r.initialCapSize ?? n.initialCapSize,
          initialCapOffsetX: r.initialCapOffsetX ?? n.initialCapOffsetX,
          initialCapOffsetY: r.initialCapOffsetY ?? n.initialCapOffsetY,
          bodyFontCSS: r.bodyFontCSS ?? n.bodyFontCSS,
          bodyLineHeight: r.bodyLineHeight ?? n.bodyLineHeight,
          headingLineHeight: r.headingLineHeight ?? n.headingLineHeight,
          breakpointIndex: s
        };
  }
  return {
    images: n.images,
    columns: n.columns,
    editorWidth: n.editorWidth,
    fontFamily: n.fontFamily,
    fontSize: n.fontSize,
    initialCap: n.initialCap,
    initialCapFont: n.initialCapFont,
    initialCapSize: n.initialCapSize,
    initialCapOffsetX: n.initialCapOffsetX,
    initialCapOffsetY: n.initialCapOffsetY,
    bodyFontCSS: n.bodyFontCSS,
    bodyLineHeight: n.bodyLineHeight,
    headingLineHeight: n.headingLineHeight,
    breakpointIndex: -1
  };
}
function Et({
  blocks: n,
  layout: e,
  config: t,
  availableFonts: r,
  availableInitialFonts: s,
  resolveImageUrl: o,
  className: l,
  style: i,
  editorMode: c,
  containerRef: p
}) {
  const f = Tn(), y = ue(null), b = p || y, [d, I] = ke(null), [z, T] = ke(!1), [J, ne] = ke(0);
  Me(() => {
    document.fonts.ready.then(() => T(!0));
    const h = () => ne((w) => w + 1);
    return document.fonts.addEventListener("loadingdone", h), () => document.fonts.removeEventListener("loadingdone", h);
  }, []);
  const B = ae(() => {
    if (!f || !b.current || n.length === 0) return;
    const h = b.current.offsetWidth, w = _n(e, h), k = w.editorWidth || 700, H = h / k, Q = En(w.images || [], H), W = w.fontFamily, A = r == null ? void 0 : r.find((te) => te.name === W), O = "Lato, sans-serif";
    let m;
    w.bodyFontCSS ? m = w.bodyFontCSS.replace(/^\d+px\s*/, "") : A ? m = A.bodyFont.replace(/^\d+px\s*/, "") : m = O;
    const S = w.fontSize || 16, P = w.bodyLineHeight ?? (A == null ? void 0 : A.bodyLineHeight) ?? Math.round(S * 1.6), g = w.headingLineHeight ?? (A == null ? void 0 : A.headingLineHeight) ?? Math.round(S * 1.5 * 1.4), X = {
      bodyFont: `${S}px ${m}`,
      headingFont: `bold ${S}px ${m}`,
      h1Font: `bold ${Math.round(S * 2)}px ${m}`,
      h2Font: `bold ${Math.round(S * 1.5)}px ${m}`,
      h3Font: `bold ${Math.round(S * 1.25)}px ${m}`,
      bodyLineHeight: P,
      h1LineHeight: Math.round(S * 2 * 1.3),
      h2LineHeight: g,
      h3LineHeight: Math.round(S * 1.25 * 1.4)
    }, C = w.initialCap || !1, V = s == null ? void 0 : s.find((te) => te.name === w.initialCapFont), M = V ? `${w.initialCapSize || 96}px ${V.fontFamily}` : e.initialCapFontCSS || `${w.initialCapSize || 96}px serif`, j = {
      ...t,
      ...X,
      dropCap: C,
      dropCapFont: M,
      dropCapOffsetX: w.initialCapOffsetX ?? 0,
      dropCapOffsetY: w.initialCapOffsetY ?? 0,
      columns: w.columns || (t == null ? void 0 : t.columns) || 1
    }, ie = (te, Z) => {
      const ce = document.createElement("span");
      ce.textContent = te, ce.style.cssText = `position:absolute;font:${Z};visibility:hidden;line-height:1;`, b.current.appendChild(ce);
      const me = ce.getBoundingClientRect();
      return b.current.removeChild(ce), { width: me.width, height: me.height };
    };
    if (C && M && !document.fonts.check(M)) {
      document.fonts.load(M).then(() => B());
      return;
    }
    const ee = Hn(n, Q, h, f, j, ie);
    I({ ...ee, activeImages: w.images || [] });
  }, [f, n, e, t, r, s, z, J]);
  Me(() => {
    B();
  }, [B]), Me(() => {
    if (!b.current) return;
    const h = new ResizeObserver(() => B());
    return h.observe(b.current), () => h.disconnect();
  }, [B]);
  const E = (h, w) => o ? o(h, w) : h, $ = !!c, D = (h, w, k) => {
    if (!w || w.length === 0) return h;
    const H = k || 0, Q = H + h.length;
    let W = 0;
    const A = [];
    let O = 0;
    for (const m of w) {
      const S = W + m.text.length, P = Math.max(W, H), g = Math.min(S, Q);
      if (P < g) {
        const X = m.text.slice(P - W, g - W), C = {};
        m.bold && (C.fontWeight = "bold"), m.italic && (C.fontStyle = "italic"), m.code && (C.fontFamily = "'SF Mono', 'Fira Code', Consolas, monospace", C.background = "rgba(80,37,129,0.08)", C.padding = "1px 4px", C.borderRadius = "3px", C.fontSize = "0.9em"), m.link && (C.color = "#502581", C.textDecoration = "underline"), m.strikethrough && (C.textDecoration = "line-through"), m.link && !$ ? A.push(
          /* @__PURE__ */ v(
            "a",
            {
              href: m.link,
              target: "_blank",
              rel: "noopener noreferrer",
              style: C,
              children: X
            },
            O++
          )
        ) : Object.keys(C).length > 0 ? A.push(/* @__PURE__ */ v("span", { style: C, children: X }, O++)) : A.push(/* @__PURE__ */ v(Ke.Fragment, { children: X }, O++));
      }
      W = S;
    }
    return A.length > 0 ? A : h;
  };
  return /* @__PURE__ */ K(
    "div",
    {
      ref: b,
      className: l,
      style: { position: "relative", minHeight: "100%", ...i },
      onMouseDown: c == null ? void 0 : c.onBackgroundMouseDown,
      onClick: c == null ? void 0 : c.onBackgroundClick,
      onMouseMove: c == null ? void 0 : c.onMouseMove,
      onMouseUp: c == null ? void 0 : c.onMouseUp,
      onMouseLeave: c == null ? void 0 : c.onMouseLeave,
      children: [
        !d && /* @__PURE__ */ v("div", { children: n.map((h, w) => h.type === "heading" ? /* @__PURE__ */ v("h2", { children: h.text }, w) : /* @__PURE__ */ v("p", { children: h.text }, w)) }),
        d && /* @__PURE__ */ K("div", { style: { position: "relative", height: d.totalHeight }, children: [
          d.elements.map((h, w) => {
            if (h.type === "dropCap")
              return /* @__PURE__ */ v(
                "span",
                {
                  style: {
                    position: "absolute",
                    font: h.font,
                    color: "#502581",
                    left: h.x,
                    top: h.y,
                    lineHeight: 1,
                    pointerEvents: $ ? "none" : void 0
                  },
                  children: h.char
                },
                `dropcap-${w}`
              );
            if (h.type === "text")
              return /* @__PURE__ */ v(
                "span",
                {
                  style: {
                    position: "absolute",
                    font: h.font,
                    whiteSpace: "pre",
                    left: h.x,
                    top: h.y,
                    color: "#333",
                    pointerEvents: $ ? "none" : void 0
                  },
                  children: D(h.text || "", h.segments, h.charOffset || 0)
                },
                `text-${w}`
              );
            if (h.type === "listBullet")
              return /* @__PURE__ */ v(
                "span",
                {
                  style: {
                    position: "absolute",
                    font: h.font,
                    left: h.x,
                    top: h.y,
                    color: "#502581",
                    pointerEvents: $ ? "none" : void 0,
                    fontWeight: 600
                  },
                  children: h.text
                },
                `bullet-${w}`
              );
            if (h.type === "blockquoteBorder")
              return /* @__PURE__ */ v(
                "div",
                {
                  style: {
                    position: "absolute",
                    left: h.x,
                    top: h.y,
                    width: h.width,
                    height: h.height,
                    background: "#502581",
                    borderRadius: 2,
                    opacity: 0.4,
                    pointerEvents: $ ? "none" : void 0
                  }
                },
                `bqborder-${w}`
              );
            if (h.type === "codeBlock")
              return /* @__PURE__ */ v(
                "div",
                {
                  style: {
                    position: "absolute",
                    left: h.x,
                    top: h.y,
                    width: h.width,
                    height: h.height,
                    background: "#f5f2f0",
                    border: "1px solid #e0d8d0",
                    borderRadius: 4,
                    pointerEvents: $ ? "none" : void 0
                  }
                },
                `code-${w}`
              );
            if (h.type === "hr")
              return /* @__PURE__ */ v(
                "div",
                {
                  style: {
                    position: "absolute",
                    left: h.x,
                    top: h.y,
                    width: h.width,
                    height: 1,
                    background: "#ccc",
                    pointerEvents: $ ? "none" : void 0
                  }
                },
                `hr-${w}`
              );
            if (h.type === "image") {
              const k = h.imageIndex, H = d.activeImages[k], Q = H ? E(H.url, H.filename) : h.url, W = $ && c.selectedImageIndex === k, A = $ && c.drawingPolygonIndex === k, O = h.polygon && h.polygon.length >= 3, m = O ? `polygon(${h.polygon.map((P) => `${P.x * 100}% ${P.y * 100}%`).join(", ")})` : void 0, S = $ ? A ? "crosshair" : "grab" : void 0;
              return /* @__PURE__ */ K(Ke.Fragment, { children: [
                $ && O && /* @__PURE__ */ v(
                  "img",
                  {
                    src: Q,
                    alt: "",
                    "data-image-index": k,
                    onMouseDown: (P) => {
                      var g;
                      return (g = c.onImageMouseDown) == null ? void 0 : g.call(c, P, k);
                    },
                    style: {
                      position: "absolute",
                      left: h.x,
                      top: h.y,
                      width: h.width,
                      borderRadius: 4,
                      opacity: 0.15,
                      cursor: S
                    }
                  }
                ),
                /* @__PURE__ */ v(
                  "img",
                  {
                    src: Q,
                    alt: h.alt,
                    "data-image-index": k,
                    onMouseDown: $ ? (P) => {
                      var g;
                      return (g = c.onImageMouseDown) == null ? void 0 : g.call(c, P, k);
                    } : void 0,
                    style: {
                      position: "absolute",
                      left: h.x,
                      top: h.y,
                      width: h.width,
                      border: $ ? `2px solid ${W ? "#502581" : "transparent"}` : "none",
                      borderRadius: 4,
                      clipPath: m,
                      cursor: S,
                      pointerEvents: $ && O ? "none" : void 0
                    }
                  }
                )
              ] }, `img-${w}`);
            }
            return null;
          }),
          $ && d.elements.map((h, w) => {
            var O;
            if (h.type !== "image") return null;
            const k = h.imageIndex, H = h.polygon || [], Q = c.selectedImageIndex === k, W = c.drawingPolygonIndex === k, A = h.width * (((O = d.activeImages[k]) == null ? void 0 : O.aspectRatio) || 1.2);
            return /* @__PURE__ */ K(Ke.Fragment, { children: [
              (H.length > 0 || W) && /* @__PURE__ */ K(
                "svg",
                {
                  style: {
                    position: "absolute",
                    left: h.x,
                    top: h.y,
                    width: h.width,
                    height: A,
                    pointerEvents: "none"
                  },
                  viewBox: "0 0 1 1",
                  preserveAspectRatio: "none",
                  children: [
                    H.length >= 2 && /* @__PURE__ */ v(
                      "path",
                      {
                        d: H.map((m, S) => `${S === 0 ? "M" : "L"} ${m.x} ${m.y}`).join(" ") + (H.length >= 3 ? " Z" : ""),
                        fill: "rgba(80,37,129,0.1)",
                        stroke: "#502581",
                        strokeWidth: "0.008",
                        strokeDasharray: "0.02 0.015"
                      }
                    ),
                    H.map((m, S) => /* @__PURE__ */ v(
                      "circle",
                      {
                        cx: m.x,
                        cy: m.y,
                        r: W ? 0.025 : 0.015,
                        fill: W ? "#502581" : "rgba(80,37,129,0.5)"
                      },
                      S
                    ))
                  ]
                }
              ),
              Q && /* @__PURE__ */ v(
                "div",
                {
                  "data-resize-handle": k,
                  onMouseDown: (m) => {
                    var S;
                    return (S = c.onResizeMouseDown) == null ? void 0 : S.call(c, m, k);
                  },
                  style: {
                    position: "absolute",
                    left: h.x + h.width - 4,
                    top: h.y + A - 4,
                    width: 8,
                    height: 8,
                    background: "#502581",
                    cursor: "nwse-resize",
                    borderRadius: 1,
                    opacity: 0.7
                  }
                }
              )
            ] }, `editor-${w}`);
          })
        ] })
      ]
    }
  );
}
function it() {
  return { async: !1, breaks: !1, extensions: null, gfm: !0, hooks: null, pedantic: !1, renderer: null, silent: !1, tokenizer: null, walkTokens: null };
}
var ze = it();
function Ft(n) {
  ze = n;
}
var Ce = { exec: () => null };
function F(n, e = "") {
  let t = typeof n == "string" ? n : n.source, r = { replace: (s, o) => {
    let l = typeof o == "string" ? o : o.source;
    return l = l.replace(se.caret, "$1"), t = t.replace(s, l), r;
  }, getRegex: () => new RegExp(t, e) };
  return r;
}
var Bn = (() => {
  try {
    return !!new RegExp("(?<=1)(?<!1)");
  } catch {
    return !1;
  }
})(), se = { codeRemoveIndent: /^(?: {1,4}| {0,3}\t)/gm, outputLinkReplace: /\\([\[\]])/g, indentCodeCompensation: /^(\s+)(?:```)/, beginningSpace: /^\s+/, endingHash: /#$/, startingSpaceChar: /^ /, endingSpaceChar: / $/, nonSpaceChar: /[^ ]/, newLineCharGlobal: /\n/g, tabCharGlobal: /\t/g, multipleSpaceGlobal: /\s+/g, blankLine: /^[ \t]*$/, doubleBlankLine: /\n[ \t]*\n[ \t]*$/, blockquoteStart: /^ {0,3}>/, blockquoteSetextReplace: /\n {0,3}((?:=+|-+) *)(?=\n|$)/g, blockquoteSetextReplace2: /^ {0,3}>[ \t]?/gm, listReplaceNesting: /^ {1,4}(?=( {4})*[^ ])/g, listIsTask: /^\[[ xX]\] +\S/, listReplaceTask: /^\[[ xX]\] +/, listTaskCheckbox: /\[[ xX]\]/, anyLine: /\n.*\n/, hrefBrackets: /^<(.*)>$/, tableDelimiter: /[:|]/, tableAlignChars: /^\||\| *$/g, tableRowBlankLine: /\n[ \t]*$/, tableAlignRight: /^ *-+: *$/, tableAlignCenter: /^ *:-+: *$/, tableAlignLeft: /^ *:-+ *$/, startATag: /^<a /i, endATag: /^<\/a>/i, startPreScriptTag: /^<(pre|code|kbd|script)(\s|>)/i, endPreScriptTag: /^<\/(pre|code|kbd|script)(\s|>)/i, startAngleBracket: /^</, endAngleBracket: />$/, pedanticHrefTitle: /^([^'"]*[^\s])\s+(['"])(.*)\2/, unicodeAlphaNumeric: /[\p{L}\p{N}]/u, escapeTest: /[&<>"']/, escapeReplace: /[&<>"']/g, escapeTestNoEncode: /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/, escapeReplaceNoEncode: /[<>"']|&(?!(#\d{1,7}|#[Xx][a-fA-F0-9]{1,6}|\w+);)/g, caret: /(^|[^\[])\^/g, percentDecode: /%25/g, findPipe: /\|/g, splitPipe: / \|/, slashPipe: /\\\|/g, carriageReturn: /\r\n|\r/g, spaceLine: /^ +$/gm, notSpaceStart: /^\S*/, endingNewline: /\n$/, listItemRegex: (n) => new RegExp(`^( {0,3}${n})((?:[	 ][^\\n]*)?(?:\\n|$))`), nextBulletRegex: (n) => new RegExp(`^ {0,${Math.min(3, n - 1)}}(?:[*+-]|\\d{1,9}[.)])((?:[ 	][^\\n]*)?(?:\\n|$))`), hrRegex: (n) => new RegExp(`^ {0,${Math.min(3, n - 1)}}((?:- *){3,}|(?:_ *){3,}|(?:\\* *){3,})(?:\\n+|$)`), fencesBeginRegex: (n) => new RegExp(`^ {0,${Math.min(3, n - 1)}}(?:\`\`\`|~~~)`), headingBeginRegex: (n) => new RegExp(`^ {0,${Math.min(3, n - 1)}}#`), htmlBeginRegex: (n) => new RegExp(`^ {0,${Math.min(3, n - 1)}}<(?:[a-z].*>|!--)`, "i"), blockquoteBeginRegex: (n) => new RegExp(`^ {0,${Math.min(3, n - 1)}}>`) }, Dn = /^(?:[ \t]*(?:\n|$))+/, On = /^((?: {4}| {0,3}\t)[^\n]+(?:\n(?:[ \t]*(?:\n|$))*)?)+/, qn = /^ {0,3}(`{3,}(?=[^`\n]*(?:\n|$))|~{3,})([^\n]*)(?:\n|$)(?:|([\s\S]*?)(?:\n|$))(?: {0,3}\1[~`]* *(?=\n|$)|$)/, He = /^ {0,3}((?:-[\t ]*){3,}|(?:_[ \t]*){3,}|(?:\*[ \t]*){3,})(?:\n+|$)/, Wn = /^ {0,3}(#{1,6})(?=\s|$)(.*)(?:\n+|$)/, ot = / {0,3}(?:[*+-]|\d{1,9}[.)])/, Ht = /^(?!bull |blockCode|fences|blockquote|heading|html|table)((?:.|\n(?!\s*?\n|bull |blockCode|fences|blockquote|heading|html|table))+?)\n {0,3}(=+|-+) *(?:\n+|$)/, _t = F(Ht).replace(/bull/g, ot).replace(/blockCode/g, /(?: {4}| {0,3}\t)/).replace(/fences/g, / {0,3}(?:`{3,}|~{3,})/).replace(/blockquote/g, / {0,3}>/).replace(/heading/g, / {0,3}#{1,6}/).replace(/html/g, / {0,3}<[^\n>]+>\n/).replace(/\|table/g, "").getRegex(), Xn = F(Ht).replace(/bull/g, ot).replace(/blockCode/g, /(?: {4}| {0,3}\t)/).replace(/fences/g, / {0,3}(?:`{3,}|~{3,})/).replace(/blockquote/g, / {0,3}>/).replace(/heading/g, / {0,3}#{1,6}/).replace(/html/g, / {0,3}<[^\n>]+>\n/).replace(/table/g, / {0,3}\|?(?:[:\- ]*\|)+[\:\- ]*\n/).getRegex(), lt = /^([^\n]+(?:\n(?!hr|heading|lheading|blockquote|fences|list|html|table| +\n)[^\n]+)*)/, Nn = /^[^\n]+/, at = /(?!\s*\])(?:\\[\s\S]|[^\[\]\\])+/, Yn = F(/^ {0,3}\[(label)\]: *(?:\n[ \t]*)?([^<\s][^\s]*|<.*?>)(?:(?: +(?:\n[ \t]*)?| *\n[ \t]*)(title))? *(?:\n+|$)/).replace("label", at).replace("title", /(?:"(?:\\"?|[^"\\])*"|'[^'\n]*(?:\n[^'\n]+)*\n?'|\([^()]*\))/).getRegex(), jn = F(/^(bull)([ \t][^\n]+?)?(?:\n|$)/).replace(/bull/g, ot).getRegex(), je = "address|article|aside|base|basefont|blockquote|body|caption|center|col|colgroup|dd|details|dialog|dir|div|dl|dt|fieldset|figcaption|figure|footer|form|frame|frameset|h[1-6]|head|header|hr|html|iframe|legend|li|link|main|menu|menuitem|meta|nav|noframes|ol|optgroup|option|p|param|search|section|summary|table|tbody|td|tfoot|th|thead|title|tr|track|ul", ct = /<!--(?:-?>|[\s\S]*?(?:-->|$))/, Zn = F("^ {0,3}(?:<(script|pre|style|textarea)[\\s>][\\s\\S]*?(?:</\\1>[^\\n]*\\n+|$)|comment[^\\n]*(\\n+|$)|<\\?[\\s\\S]*?(?:\\?>\\n*|$)|<![A-Z][\\s\\S]*?(?:>\\n*|$)|<!\\[CDATA\\[[\\s\\S]*?(?:\\]\\]>\\n*|$)|</?(tag)(?: +|\\n|/?>)[\\s\\S]*?(?:(?:\\n[ 	]*)+\\n|$)|<(?!script|pre|style|textarea)([a-z][\\w-]*)(?:attribute)*? */?>(?=[ \\t]*(?:\\n|$))[\\s\\S]*?(?:(?:\\n[ 	]*)+\\n|$)|</(?!script|pre|style|textarea)[a-z][\\w-]*\\s*>(?=[ \\t]*(?:\\n|$))[\\s\\S]*?(?:(?:\\n[ 	]*)+\\n|$))", "i").replace("comment", ct).replace("tag", je).replace("attribute", / +[a-zA-Z:_][\w.:-]*(?: *= *"[^"\n]*"| *= *'[^'\n]*'| *= *[^\s"'=<>`]+)?/).getRegex(), Bt = F(lt).replace("hr", He).replace("heading", " {0,3}#{1,6}(?:\\s|$)").replace("|lheading", "").replace("|table", "").replace("blockquote", " {0,3}>").replace("fences", " {0,3}(?:`{3,}(?=[^`\\n]*\\n)|~{3,})[^\\n]*\\n").replace("list", " {0,3}(?:[*+-]|1[.)])[ \\t]").replace("html", "</?(?:tag)(?: +|\\n|/?>)|<(?:script|pre|style|textarea|!--)").replace("tag", je).getRegex(), Gn = F(/^( {0,3}> ?(paragraph|[^\n]*)(?:\n|$))+/).replace("paragraph", Bt).getRegex(), pt = { blockquote: Gn, code: On, def: Yn, fences: qn, heading: Wn, hr: He, html: Zn, lheading: _t, list: jn, newline: Dn, paragraph: Bt, table: Ce, text: Nn }, Rt = F("^ *([^\\n ].*)\\n {0,3}((?:\\| *)?:?-+:? *(?:\\| *:?-+:? *)*(?:\\| *)?)(?:\\n((?:(?! *\\n|hr|heading|blockquote|code|fences|list|html).*(?:\\n|$))*)\\n*|$)").replace("hr", He).replace("heading", " {0,3}#{1,6}(?:\\s|$)").replace("blockquote", " {0,3}>").replace("code", "(?: {4}| {0,3}	)[^\\n]").replace("fences", " {0,3}(?:`{3,}(?=[^`\\n]*\\n)|~{3,})[^\\n]*\\n").replace("list", " {0,3}(?:[*+-]|1[.)])[ \\t]").replace("html", "</?(?:tag)(?: +|\\n|/?>)|<(?:script|pre|style|textarea|!--)").replace("tag", je).getRegex(), Qn = { ...pt, lheading: Xn, table: Rt, paragraph: F(lt).replace("hr", He).replace("heading", " {0,3}#{1,6}(?:\\s|$)").replace("|lheading", "").replace("table", Rt).replace("blockquote", " {0,3}>").replace("fences", " {0,3}(?:`{3,}(?=[^`\\n]*\\n)|~{3,})[^\\n]*\\n").replace("list", " {0,3}(?:[*+-]|1[.)])[ \\t]").replace("html", "</?(?:tag)(?: +|\\n|/?>)|<(?:script|pre|style|textarea|!--)").replace("tag", je).getRegex() }, Un = { ...pt, html: F(`^ *(?:comment *(?:\\n|\\s*$)|<(tag)[\\s\\S]+?</\\1> *(?:\\n{2,}|\\s*$)|<tag(?:"[^"]*"|'[^']*'|\\s[^'"/>\\s]*)*?/?> *(?:\\n{2,}|\\s*$))`).replace("comment", ct).replace(/tag/g, "(?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img)\\b)\\w+(?!:|[^\\w\\s@]*@)\\b").getRegex(), def: /^ *\[([^\]]+)\]: *<?([^\s>]+)>?(?: +(["(][^\n]+[")]))? *(?:\n+|$)/, heading: /^(#{1,6})(.*)(?:\n+|$)/, fences: Ce, lheading: /^(.+?)\n {0,3}(=+|-+) *(?:\n+|$)/, paragraph: F(lt).replace("hr", He).replace("heading", ` *#{1,6} *[^
]`).replace("lheading", _t).replace("|table", "").replace("blockquote", " {0,3}>").replace("|fences", "").replace("|list", "").replace("|html", "").replace("|tag", "").getRegex() }, Kn = /^\\([!"#$%&'()*+,\-./:;<=>?@\[\]\\^_`{|}~])/, Vn = /^(`+)([^`]|[^`][\s\S]*?[^`])\1(?!`)/, Dt = /^( {2,}|\\)\n(?!\s*$)/, Jn = /^(`+|[^`])(?:(?= {2,}\n)|[\s\S]*?(?:(?=[\\<!\[`*_]|\b_|$)|[^ ](?= {2,}\n)))/, Pe = /[\p{P}\p{S}]/u, Ze = /[\s\p{P}\p{S}]/u, ht = /[^\s\p{P}\p{S}]/u, er = F(/^((?![*_])punctSpace)/, "u").replace(/punctSpace/g, Ze).getRegex(), Ot = /(?!~)[\p{P}\p{S}]/u, tr = /(?!~)[\s\p{P}\p{S}]/u, nr = /(?:[^\s\p{P}\p{S}]|~)/u, rr = F(/link|precode-code|html/, "g").replace("link", /\[(?:[^\[\]`]|(?<a>`+)[^`]+\k<a>(?!`))*?\]\((?:\\[\s\S]|[^\\\(\)]|\((?:\\[\s\S]|[^\\\(\)])*\))*\)/).replace("precode-", Bn ? "(?<!`)()" : "(^^|[^`])").replace("code", /(?<b>`+)[^`]+\k<b>(?!`)/).replace("html", /<(?! )[^<>]*?>/).getRegex(), qt = /^(?:\*+(?:((?!\*)punct)|([^\s*]))?)|^_+(?:((?!_)punct)|([^\s_]))?/, sr = F(qt, "u").replace(/punct/g, Pe).getRegex(), ir = F(qt, "u").replace(/punct/g, Ot).getRegex(), Wt = "^[^_*]*?__[^_*]*?\\*[^_*]*?(?=__)|[^*]+(?=[^*])|(?!\\*)punct(\\*+)(?=[\\s]|$)|notPunctSpace(\\*+)(?!\\*)(?=punctSpace|$)|(?!\\*)punctSpace(\\*+)(?=notPunctSpace)|[\\s](\\*+)(?!\\*)(?=punct)|(?!\\*)punct(\\*+)(?!\\*)(?=punct)|notPunctSpace(\\*+)(?=notPunctSpace)", or = F(Wt, "gu").replace(/notPunctSpace/g, ht).replace(/punctSpace/g, Ze).replace(/punct/g, Pe).getRegex(), lr = F(Wt, "gu").replace(/notPunctSpace/g, nr).replace(/punctSpace/g, tr).replace(/punct/g, Ot).getRegex(), ar = F("^[^_*]*?\\*\\*[^_*]*?_[^_*]*?(?=\\*\\*)|[^_]+(?=[^_])|(?!_)punct(_+)(?=[\\s]|$)|notPunctSpace(_+)(?!_)(?=punctSpace|$)|(?!_)punctSpace(_+)(?=notPunctSpace)|[\\s](_+)(?!_)(?=punct)|(?!_)punct(_+)(?!_)(?=punct)", "gu").replace(/notPunctSpace/g, ht).replace(/punctSpace/g, Ze).replace(/punct/g, Pe).getRegex(), cr = F(/^~~?(?:((?!~)punct)|[^\s~])/, "u").replace(/punct/g, Pe).getRegex(), pr = "^[^~]+(?=[^~])|(?!~)punct(~~?)(?=[\\s]|$)|notPunctSpace(~~?)(?!~)(?=punctSpace|$)|(?!~)punctSpace(~~?)(?=notPunctSpace)|[\\s](~~?)(?!~)(?=punct)|(?!~)punct(~~?)(?!~)(?=punct)|notPunctSpace(~~?)(?=notPunctSpace)", hr = F(pr, "gu").replace(/notPunctSpace/g, ht).replace(/punctSpace/g, Ze).replace(/punct/g, Pe).getRegex(), ur = F(/\\(punct)/, "gu").replace(/punct/g, Pe).getRegex(), dr = F(/^<(scheme:[^\s\x00-\x1f<>]*|email)>/).replace("scheme", /[a-zA-Z][a-zA-Z0-9+.-]{1,31}/).replace("email", /[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+(@)[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)+(?![-_])/).getRegex(), gr = F(ct).replace("(?:-->|$)", "-->").getRegex(), fr = F("^comment|^</[a-zA-Z][\\w:-]*\\s*>|^<[a-zA-Z][\\w-]*(?:attribute)*?\\s*/?>|^<\\?[\\s\\S]*?\\?>|^<![a-zA-Z]+\\s[\\s\\S]*?>|^<!\\[CDATA\\[[\\s\\S]*?\\]\\]>").replace("comment", gr).replace("attribute", /\s+[a-zA-Z:_][\w.:-]*(?:\s*=\s*"[^"]*"|\s*=\s*'[^']*'|\s*=\s*[^\s"'=<>`]+)?/).getRegex(), Xe = /(?:\[(?:\\[\s\S]|[^\[\]\\])*\]|\\[\s\S]|`+(?!`)[^`]*?`+(?!`)|``+(?=\])|[^\[\]\\`])*?/, xr = F(/^!?\[(label)\]\(\s*(href)(?:(?:[ \t]+(?:\n[ \t]*)?|\n[ \t]*)(title))?\s*\)/).replace("label", Xe).replace("href", /<(?:\\.|[^\n<>\\])+>|[^ \t\n\x00-\x1f]*/).replace("title", /"(?:\\"?|[^"\\])*"|'(?:\\'?|[^'\\])*'|\((?:\\\)?|[^)\\])*\)/).getRegex(), Xt = F(/^!?\[(label)\]\[(ref)\]/).replace("label", Xe).replace("ref", at).getRegex(), Nt = F(/^!?\[(ref)\](?:\[\])?/).replace("ref", at).getRegex(), kr = F("reflink|nolink(?!\\()", "g").replace("reflink", Xt).replace("nolink", Nt).getRegex(), $t = /[hH][tT][tT][pP][sS]?|[fF][tT][pP]/, ut = { _backpedal: Ce, anyPunctuation: ur, autolink: dr, blockSkip: rr, br: Dt, code: Vn, del: Ce, delLDelim: Ce, delRDelim: Ce, emStrongLDelim: sr, emStrongRDelimAst: or, emStrongRDelimUnd: ar, escape: Kn, link: xr, nolink: Nt, punctuation: er, reflink: Xt, reflinkSearch: kr, tag: fr, text: Jn, url: Ce }, mr = { ...ut, link: F(/^!?\[(label)\]\((.*?)\)/).replace("label", Xe).getRegex(), reflink: F(/^!?\[(label)\]\s*\[([^\]]*)\]/).replace("label", Xe).getRegex() }, et = { ...ut, emStrongRDelimAst: lr, emStrongLDelim: ir, delLDelim: cr, delRDelim: hr, url: F(/^((?:protocol):\/\/|www\.)(?:[a-zA-Z0-9\-]+\.?)+[^\s<]*|^email/).replace("protocol", $t).replace("email", /[A-Za-z0-9._+-]+(@)[a-zA-Z0-9-_]+(?:\.[a-zA-Z0-9-_]*[a-zA-Z0-9])+(?![-_])/).getRegex(), _backpedal: /(?:[^?!.,:;*_'"~()&]+|\([^)]*\)|&(?![a-zA-Z0-9]+;$)|[?!.,:;*_'"~)]+(?!$))+/, del: /^(~~?)(?=[^\s~])((?:\\[\s\S]|[^\\])*?(?:\\[\s\S]|[^\s~\\]))\1(?=[^~]|$)/, text: F(/^([`~]+|[^`~])(?:(?= {2,}\n)|(?=[a-zA-Z0-9.!#$%&'*+\/=?_`{\|}~-]+@)|[\s\S]*?(?:(?=[\\<!\[`*~_]|\b_|protocol:\/\/|www\.|$)|[^ ](?= {2,}\n)|[^a-zA-Z0-9.!#$%&'*+\/=?_`{\|}~-](?=[a-zA-Z0-9.!#$%&'*+\/=?_`{\|}~-]+@)))/).replace("protocol", $t).getRegex() }, br = { ...et, br: F(Dt).replace("{2,}", "*").getRegex(), text: F(et.text).replace("\\b_", "\\b_| {2,}\\n").replace(/\{2,\}/g, "*").getRegex() }, De = { normal: pt, gfm: Qn, pedantic: Un }, Ee = { normal: ut, gfm: et, breaks: br, pedantic: mr }, yr = { "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }, zt = (n) => yr[n];
function xe(n, e) {
  if (e) {
    if (se.escapeTest.test(n)) return n.replace(se.escapeReplace, zt);
  } else if (se.escapeTestNoEncode.test(n)) return n.replace(se.escapeReplaceNoEncode, zt);
  return n;
}
function vt(n) {
  try {
    n = encodeURI(n).replace(se.percentDecode, "%");
  } catch {
    return null;
  }
  return n;
}
function It(n, e) {
  var o;
  let t = n.replace(se.findPipe, (l, i, c) => {
    let p = !1, f = i;
    for (; --f >= 0 && c[f] === "\\"; ) p = !p;
    return p ? "|" : " |";
  }), r = t.split(se.splitPipe), s = 0;
  if (r[0].trim() || r.shift(), r.length > 0 && !((o = r.at(-1)) != null && o.trim()) && r.pop(), e) if (r.length > e) r.splice(e);
  else for (; r.length < e; ) r.push("");
  for (; s < r.length; s++) r[s] = r[s].trim().replace(se.slashPipe, "|");
  return r;
}
function ye(n, e, t) {
  let r = n.length;
  if (r === 0) return "";
  let s = 0;
  for (; s < r && n.charAt(r - s - 1) === e; )
    s++;
  return n.slice(0, r - s);
}
function Lt(n) {
  let e = n.split(`
`), t = e.length - 1;
  for (; t >= 0 && !e[t].trim(); ) t--;
  return e.length - t <= 2 ? n : e.slice(0, t + 1).join(`
`);
}
function wr(n, e) {
  if (n.indexOf(e[1]) === -1) return -1;
  let t = 0;
  for (let r = 0; r < n.length; r++) if (n[r] === "\\") r++;
  else if (n[r] === e[0]) t++;
  else if (n[r] === e[1] && (t--, t < 0)) return r;
  return t > 0 ? -2 : -1;
}
function Sr(n, e = 0) {
  let t = e, r = "";
  for (let s of n) if (s === "	") {
    let o = 4 - t % 4;
    r += " ".repeat(o), t += o;
  } else r += s, t++;
  return r;
}
function Tt(n, e, t, r, s) {
  let o = e.href, l = e.title || null, i = n[1].replace(s.other.outputLinkReplace, "$1");
  r.state.inLink = !0;
  let c = { type: n[0].charAt(0) === "!" ? "image" : "link", raw: t, href: o, title: l, text: i, tokens: r.inlineTokens(i) };
  return r.state.inLink = !1, c;
}
function Cr(n, e, t) {
  let r = n.match(t.other.indentCodeCompensation);
  if (r === null) return e;
  let s = r[1];
  return e.split(`
`).map((o) => {
    let l = o.match(t.other.beginningSpace);
    if (l === null) return o;
    let [i] = l;
    return i.length >= s.length ? o.slice(s.length) : o;
  }).join(`
`);
}
var Ne = class {
  constructor(n) {
    Y(this, "options");
    Y(this, "rules");
    Y(this, "lexer");
    this.options = n || ze;
  }
  space(n) {
    let e = this.rules.block.newline.exec(n);
    if (e && e[0].length > 0) return { type: "space", raw: e[0] };
  }
  code(n) {
    let e = this.rules.block.code.exec(n);
    if (e) {
      let t = this.options.pedantic ? e[0] : Lt(e[0]), r = t.replace(this.rules.other.codeRemoveIndent, "");
      return { type: "code", raw: t, codeBlockStyle: "indented", text: r };
    }
  }
  fences(n) {
    let e = this.rules.block.fences.exec(n);
    if (e) {
      let t = e[0], r = Cr(t, e[3] || "", this.rules);
      return { type: "code", raw: t, lang: e[2] ? e[2].trim().replace(this.rules.inline.anyPunctuation, "$1") : e[2], text: r };
    }
  }
  heading(n) {
    let e = this.rules.block.heading.exec(n);
    if (e) {
      let t = e[2].trim();
      if (this.rules.other.endingHash.test(t)) {
        let r = ye(t, "#");
        (this.options.pedantic || !r || this.rules.other.endingSpaceChar.test(r)) && (t = r.trim());
      }
      return { type: "heading", raw: ye(e[0], `
`), depth: e[1].length, text: t, tokens: this.lexer.inline(t) };
    }
  }
  hr(n) {
    let e = this.rules.block.hr.exec(n);
    if (e) return { type: "hr", raw: ye(e[0], `
`) };
  }
  blockquote(n) {
    let e = this.rules.block.blockquote.exec(n);
    if (e) {
      let t = ye(e[0], `
`).split(`
`), r = "", s = "", o = [];
      for (; t.length > 0; ) {
        let l = !1, i = [], c;
        for (c = 0; c < t.length; c++) if (this.rules.other.blockquoteStart.test(t[c])) i.push(t[c]), l = !0;
        else if (!l) i.push(t[c]);
        else break;
        t = t.slice(c);
        let p = i.join(`
`), f = p.replace(this.rules.other.blockquoteSetextReplace, `
    $1`).replace(this.rules.other.blockquoteSetextReplace2, "");
        r = r ? `${r}
${p}` : p, s = s ? `${s}
${f}` : f;
        let y = this.lexer.state.top;
        if (this.lexer.state.top = !0, this.lexer.blockTokens(f, o, !0), this.lexer.state.top = y, t.length === 0) break;
        let b = o.at(-1);
        if ((b == null ? void 0 : b.type) === "code") break;
        if ((b == null ? void 0 : b.type) === "blockquote") {
          let d = b, I = d.raw + `
` + t.join(`
`), z = this.blockquote(I);
          o[o.length - 1] = z, r = r.substring(0, r.length - d.raw.length) + z.raw, s = s.substring(0, s.length - d.text.length) + z.text;
          break;
        } else if ((b == null ? void 0 : b.type) === "list") {
          let d = b, I = d.raw + `
` + t.join(`
`), z = this.list(I);
          o[o.length - 1] = z, r = r.substring(0, r.length - b.raw.length) + z.raw, s = s.substring(0, s.length - d.raw.length) + z.raw, t = I.substring(o.at(-1).raw.length).split(`
`);
          continue;
        }
      }
      return { type: "blockquote", raw: r, tokens: o, text: s };
    }
  }
  list(n) {
    var t, r;
    let e = this.rules.block.list.exec(n);
    if (e) {
      let s = e[1].trim(), o = s.length > 1, l = { type: "list", raw: "", ordered: o, start: o ? +s.slice(0, -1) : "", loose: !1, items: [] };
      s = o ? `\\d{1,9}\\${s.slice(-1)}` : `\\${s}`, this.options.pedantic && (s = o ? s : "[*+-]");
      let i = this.rules.other.listItemRegex(s), c = !1;
      for (; n; ) {
        let f = !1, y = "", b = "";
        if (!(e = i.exec(n)) || this.rules.block.hr.test(n)) break;
        y = e[0], n = n.substring(y.length);
        let d = Sr(e[2].split(`
`, 1)[0], e[1].length), I = n.split(`
`, 1)[0], z = !d.trim(), T = 0;
        if (this.options.pedantic ? (T = 2, b = d.trimStart()) : z ? T = e[1].length + 1 : (T = d.search(this.rules.other.nonSpaceChar), T = T > 4 ? 1 : T, b = d.slice(T), T += e[1].length), z && this.rules.other.blankLine.test(I) && (y += I + `
`, n = n.substring(I.length + 1), f = !0), !f) {
          let J = this.rules.other.nextBulletRegex(T), ne = this.rules.other.hrRegex(T), B = this.rules.other.fencesBeginRegex(T), E = this.rules.other.headingBeginRegex(T), $ = this.rules.other.htmlBeginRegex(T), D = this.rules.other.blockquoteBeginRegex(T);
          for (; n; ) {
            let h = n.split(`
`, 1)[0], w;
            if (I = h, this.options.pedantic ? (I = I.replace(this.rules.other.listReplaceNesting, "  "), w = I) : w = I.replace(this.rules.other.tabCharGlobal, "    "), B.test(I) || E.test(I) || $.test(I) || D.test(I) || J.test(I) || ne.test(I)) break;
            if (w.search(this.rules.other.nonSpaceChar) >= T || !I.trim()) b += `
` + w.slice(T);
            else {
              if (z || d.replace(this.rules.other.tabCharGlobal, "    ").search(this.rules.other.nonSpaceChar) >= 4 || B.test(d) || E.test(d) || ne.test(d)) break;
              b += `
` + I;
            }
            z = !I.trim(), y += h + `
`, n = n.substring(h.length + 1), d = w.slice(T);
          }
        }
        l.loose || (c ? l.loose = !0 : this.rules.other.doubleBlankLine.test(y) && (c = !0)), l.items.push({ type: "list_item", raw: y, task: !!this.options.gfm && this.rules.other.listIsTask.test(b), loose: !1, text: b, tokens: [] }), l.raw += y;
      }
      let p = l.items.at(-1);
      if (p) p.raw = p.raw.trimEnd(), p.text = p.text.trimEnd();
      else return;
      l.raw = l.raw.trimEnd();
      for (let f of l.items) {
        if (this.lexer.state.top = !1, f.tokens = this.lexer.blockTokens(f.text, []), f.task) {
          if (f.text = f.text.replace(this.rules.other.listReplaceTask, ""), ((t = f.tokens[0]) == null ? void 0 : t.type) === "text" || ((r = f.tokens[0]) == null ? void 0 : r.type) === "paragraph") {
            f.tokens[0].raw = f.tokens[0].raw.replace(this.rules.other.listReplaceTask, ""), f.tokens[0].text = f.tokens[0].text.replace(this.rules.other.listReplaceTask, "");
            for (let b = this.lexer.inlineQueue.length - 1; b >= 0; b--) if (this.rules.other.listIsTask.test(this.lexer.inlineQueue[b].src)) {
              this.lexer.inlineQueue[b].src = this.lexer.inlineQueue[b].src.replace(this.rules.other.listReplaceTask, "");
              break;
            }
          }
          let y = this.rules.other.listTaskCheckbox.exec(f.raw);
          if (y) {
            let b = { type: "checkbox", raw: y[0] + " ", checked: y[0] !== "[ ]" };
            f.checked = b.checked, l.loose ? f.tokens[0] && ["paragraph", "text"].includes(f.tokens[0].type) && "tokens" in f.tokens[0] && f.tokens[0].tokens ? (f.tokens[0].raw = b.raw + f.tokens[0].raw, f.tokens[0].text = b.raw + f.tokens[0].text, f.tokens[0].tokens.unshift(b)) : f.tokens.unshift({ type: "paragraph", raw: b.raw, text: b.raw, tokens: [b] }) : f.tokens.unshift(b);
          }
        }
        if (!l.loose) {
          let y = f.tokens.filter((d) => d.type === "space"), b = y.length > 0 && y.some((d) => this.rules.other.anyLine.test(d.raw));
          l.loose = b;
        }
      }
      if (l.loose) for (let f of l.items) {
        f.loose = !0;
        for (let y of f.tokens) y.type === "text" && (y.type = "paragraph");
      }
      return l;
    }
  }
  html(n) {
    let e = this.rules.block.html.exec(n);
    if (e) {
      let t = Lt(e[0]);
      return { type: "html", block: !0, raw: t, pre: e[1] === "pre" || e[1] === "script" || e[1] === "style", text: t };
    }
  }
  def(n) {
    let e = this.rules.block.def.exec(n);
    if (e) {
      let t = e[1].toLowerCase().replace(this.rules.other.multipleSpaceGlobal, " "), r = e[2] ? e[2].replace(this.rules.other.hrefBrackets, "$1").replace(this.rules.inline.anyPunctuation, "$1") : "", s = e[3] ? e[3].substring(1, e[3].length - 1).replace(this.rules.inline.anyPunctuation, "$1") : e[3];
      return { type: "def", tag: t, raw: ye(e[0], `
`), href: r, title: s };
    }
  }
  table(n) {
    var l;
    let e = this.rules.block.table.exec(n);
    if (!e || !this.rules.other.tableDelimiter.test(e[2])) return;
    let t = It(e[1]), r = e[2].replace(this.rules.other.tableAlignChars, "").split("|"), s = (l = e[3]) != null && l.trim() ? e[3].replace(this.rules.other.tableRowBlankLine, "").split(`
`) : [], o = { type: "table", raw: ye(e[0], `
`), header: [], align: [], rows: [] };
    if (t.length === r.length) {
      for (let i of r) this.rules.other.tableAlignRight.test(i) ? o.align.push("right") : this.rules.other.tableAlignCenter.test(i) ? o.align.push("center") : this.rules.other.tableAlignLeft.test(i) ? o.align.push("left") : o.align.push(null);
      for (let i = 0; i < t.length; i++) o.header.push({ text: t[i], tokens: this.lexer.inline(t[i]), header: !0, align: o.align[i] });
      for (let i of s) o.rows.push(It(i, o.header.length).map((c, p) => ({ text: c, tokens: this.lexer.inline(c), header: !1, align: o.align[p] })));
      return o;
    }
  }
  lheading(n) {
    let e = this.rules.block.lheading.exec(n);
    if (e) {
      let t = e[1].trim();
      return { type: "heading", raw: ye(e[0], `
`), depth: e[2].charAt(0) === "=" ? 1 : 2, text: t, tokens: this.lexer.inline(t) };
    }
  }
  paragraph(n) {
    let e = this.rules.block.paragraph.exec(n);
    if (e) {
      let t = e[1].charAt(e[1].length - 1) === `
` ? e[1].slice(0, -1) : e[1];
      return { type: "paragraph", raw: e[0], text: t, tokens: this.lexer.inline(t) };
    }
  }
  text(n) {
    let e = this.rules.block.text.exec(n);
    if (e) return { type: "text", raw: e[0], text: e[0], tokens: this.lexer.inline(e[0]) };
  }
  escape(n) {
    let e = this.rules.inline.escape.exec(n);
    if (e) return { type: "escape", raw: e[0], text: e[1] };
  }
  tag(n) {
    let e = this.rules.inline.tag.exec(n);
    if (e) return !this.lexer.state.inLink && this.rules.other.startATag.test(e[0]) ? this.lexer.state.inLink = !0 : this.lexer.state.inLink && this.rules.other.endATag.test(e[0]) && (this.lexer.state.inLink = !1), !this.lexer.state.inRawBlock && this.rules.other.startPreScriptTag.test(e[0]) ? this.lexer.state.inRawBlock = !0 : this.lexer.state.inRawBlock && this.rules.other.endPreScriptTag.test(e[0]) && (this.lexer.state.inRawBlock = !1), { type: "html", raw: e[0], inLink: this.lexer.state.inLink, inRawBlock: this.lexer.state.inRawBlock, block: !1, text: e[0] };
  }
  link(n) {
    let e = this.rules.inline.link.exec(n);
    if (e) {
      let t = e[2].trim();
      if (!this.options.pedantic && this.rules.other.startAngleBracket.test(t)) {
        if (!this.rules.other.endAngleBracket.test(t)) return;
        let o = ye(t.slice(0, -1), "\\");
        if ((t.length - o.length) % 2 === 0) return;
      } else {
        let o = wr(e[2], "()");
        if (o === -2) return;
        if (o > -1) {
          let l = (e[0].indexOf("!") === 0 ? 5 : 4) + e[1].length + o;
          e[2] = e[2].substring(0, o), e[0] = e[0].substring(0, l).trim(), e[3] = "";
        }
      }
      let r = e[2], s = "";
      if (this.options.pedantic) {
        let o = this.rules.other.pedanticHrefTitle.exec(r);
        o && (r = o[1], s = o[3]);
      } else s = e[3] ? e[3].slice(1, -1) : "";
      return r = r.trim(), this.rules.other.startAngleBracket.test(r) && (this.options.pedantic && !this.rules.other.endAngleBracket.test(t) ? r = r.slice(1) : r = r.slice(1, -1)), Tt(e, { href: r && r.replace(this.rules.inline.anyPunctuation, "$1"), title: s && s.replace(this.rules.inline.anyPunctuation, "$1") }, e[0], this.lexer, this.rules);
    }
  }
  reflink(n, e) {
    let t;
    if ((t = this.rules.inline.reflink.exec(n)) || (t = this.rules.inline.nolink.exec(n))) {
      let r = (t[2] || t[1]).replace(this.rules.other.multipleSpaceGlobal, " "), s = e[r.toLowerCase()];
      if (!s) {
        let o = t[0].charAt(0);
        return { type: "text", raw: o, text: o };
      }
      return Tt(t, s, t[0], this.lexer, this.rules);
    }
  }
  emStrong(n, e, t = "") {
    let r = this.rules.inline.emStrongLDelim.exec(n);
    if (!(!r || !r[1] && !r[2] && !r[3] && !r[4] || r[4] && t.match(this.rules.other.unicodeAlphaNumeric)) && (!(r[1] || r[3]) || !t || this.rules.inline.punctuation.exec(t))) {
      let s = [...r[0]].length - 1, o, l, i = s, c = 0, p = r[0][0] === "*" ? this.rules.inline.emStrongRDelimAst : this.rules.inline.emStrongRDelimUnd;
      for (p.lastIndex = 0, e = e.slice(-1 * n.length + s); (r = p.exec(e)) !== null; ) {
        if (o = r[1] || r[2] || r[3] || r[4] || r[5] || r[6], !o) continue;
        if (l = [...o].length, r[3] || r[4]) {
          i += l;
          continue;
        } else if ((r[5] || r[6]) && s % 3 && !((s + l) % 3)) {
          c += l;
          continue;
        }
        if (i -= l, i > 0) continue;
        l = Math.min(l, l + i + c);
        let f = [...r[0]][0].length, y = n.slice(0, s + r.index + f + l);
        if (Math.min(s, l) % 2) {
          let d = y.slice(1, -1);
          return { type: "em", raw: y, text: d, tokens: this.lexer.inlineTokens(d) };
        }
        let b = y.slice(2, -2);
        return { type: "strong", raw: y, text: b, tokens: this.lexer.inlineTokens(b) };
      }
    }
  }
  codespan(n) {
    let e = this.rules.inline.code.exec(n);
    if (e) {
      let t = e[2].replace(this.rules.other.newLineCharGlobal, " "), r = this.rules.other.nonSpaceChar.test(t), s = this.rules.other.startingSpaceChar.test(t) && this.rules.other.endingSpaceChar.test(t);
      return r && s && (t = t.substring(1, t.length - 1)), { type: "codespan", raw: e[0], text: t };
    }
  }
  br(n) {
    let e = this.rules.inline.br.exec(n);
    if (e) return { type: "br", raw: e[0] };
  }
  del(n, e, t = "") {
    let r = this.rules.inline.delLDelim.exec(n);
    if (r && (!r[1] || !t || this.rules.inline.punctuation.exec(t))) {
      let s = [...r[0]].length - 1, o, l, i = s, c = this.rules.inline.delRDelim;
      for (c.lastIndex = 0, e = e.slice(-1 * n.length + s); (r = c.exec(e)) !== null; ) {
        if (o = r[1] || r[2] || r[3] || r[4] || r[5] || r[6], !o || (l = [...o].length, l !== s)) continue;
        if (r[3] || r[4]) {
          i += l;
          continue;
        }
        if (i -= l, i > 0) continue;
        l = Math.min(l, l + i);
        let p = [...r[0]][0].length, f = n.slice(0, s + r.index + p + l), y = f.slice(s, -s);
        return { type: "del", raw: f, text: y, tokens: this.lexer.inlineTokens(y) };
      }
    }
  }
  autolink(n) {
    let e = this.rules.inline.autolink.exec(n);
    if (e) {
      let t, r;
      return e[2] === "@" ? (t = e[1], r = "mailto:" + t) : (t = e[1], r = t), { type: "link", raw: e[0], text: t, href: r, tokens: [{ type: "text", raw: t, text: t }] };
    }
  }
  url(n) {
    var t;
    let e;
    if (e = this.rules.inline.url.exec(n)) {
      let r, s;
      if (e[2] === "@") r = e[0], s = "mailto:" + r;
      else {
        let o;
        do
          o = e[0], e[0] = ((t = this.rules.inline._backpedal.exec(e[0])) == null ? void 0 : t[0]) ?? "";
        while (o !== e[0]);
        r = e[0], e[1] === "www." ? s = "http://" + e[0] : s = e[0];
      }
      return { type: "link", raw: e[0], text: r, href: s, tokens: [{ type: "text", raw: r, text: r }] };
    }
  }
  inlineText(n) {
    let e = this.rules.inline.text.exec(n);
    if (e) {
      let t = this.lexer.state.inRawBlock;
      return { type: "text", raw: e[0], text: e[0], escaped: t };
    }
  }
}, de = class tt {
  constructor(e) {
    Y(this, "tokens");
    Y(this, "options");
    Y(this, "state");
    Y(this, "inlineQueue");
    Y(this, "tokenizer");
    this.tokens = [], this.tokens.links = /* @__PURE__ */ Object.create(null), this.options = e || ze, this.options.tokenizer = this.options.tokenizer || new Ne(), this.tokenizer = this.options.tokenizer, this.tokenizer.options = this.options, this.tokenizer.lexer = this, this.inlineQueue = [], this.state = { inLink: !1, inRawBlock: !1, top: !0 };
    let t = { other: se, block: De.normal, inline: Ee.normal };
    this.options.pedantic ? (t.block = De.pedantic, t.inline = Ee.pedantic) : this.options.gfm && (t.block = De.gfm, this.options.breaks ? t.inline = Ee.breaks : t.inline = Ee.gfm), this.tokenizer.rules = t;
  }
  static get rules() {
    return { block: De, inline: Ee };
  }
  static lex(e, t) {
    return new tt(t).lex(e);
  }
  static lexInline(e, t) {
    return new tt(t).inlineTokens(e);
  }
  lex(e) {
    e = e.replace(se.carriageReturn, `
`), this.blockTokens(e, this.tokens);
    for (let t = 0; t < this.inlineQueue.length; t++) {
      let r = this.inlineQueue[t];
      this.inlineTokens(r.src, r.tokens);
    }
    return this.inlineQueue = [], this.tokens;
  }
  blockTokens(e, t = [], r = !1) {
    var s, o, l;
    for (this.tokenizer.lexer = this, this.options.pedantic && (e = e.replace(se.tabCharGlobal, "    ").replace(se.spaceLine, "")); e; ) {
      let i;
      if ((o = (s = this.options.extensions) == null ? void 0 : s.block) != null && o.some((p) => (i = p.call({ lexer: this }, e, t)) ? (e = e.substring(i.raw.length), t.push(i), !0) : !1)) continue;
      if (i = this.tokenizer.space(e)) {
        e = e.substring(i.raw.length);
        let p = t.at(-1);
        i.raw.length === 1 && p !== void 0 ? p.raw += `
` : t.push(i);
        continue;
      }
      if (i = this.tokenizer.code(e)) {
        e = e.substring(i.raw.length);
        let p = t.at(-1);
        (p == null ? void 0 : p.type) === "paragraph" || (p == null ? void 0 : p.type) === "text" ? (p.raw += (p.raw.endsWith(`
`) ? "" : `
`) + i.raw, p.text += `
` + i.text, this.inlineQueue.at(-1).src = p.text) : t.push(i);
        continue;
      }
      if (i = this.tokenizer.fences(e)) {
        e = e.substring(i.raw.length), t.push(i);
        continue;
      }
      if (i = this.tokenizer.heading(e)) {
        e = e.substring(i.raw.length), t.push(i);
        continue;
      }
      if (i = this.tokenizer.hr(e)) {
        e = e.substring(i.raw.length), t.push(i);
        continue;
      }
      if (i = this.tokenizer.blockquote(e)) {
        e = e.substring(i.raw.length), t.push(i);
        continue;
      }
      if (i = this.tokenizer.list(e)) {
        e = e.substring(i.raw.length), t.push(i);
        continue;
      }
      if (i = this.tokenizer.html(e)) {
        e = e.substring(i.raw.length), t.push(i);
        continue;
      }
      if (i = this.tokenizer.def(e)) {
        e = e.substring(i.raw.length);
        let p = t.at(-1);
        (p == null ? void 0 : p.type) === "paragraph" || (p == null ? void 0 : p.type) === "text" ? (p.raw += (p.raw.endsWith(`
`) ? "" : `
`) + i.raw, p.text += `
` + i.raw, this.inlineQueue.at(-1).src = p.text) : this.tokens.links[i.tag] || (this.tokens.links[i.tag] = { href: i.href, title: i.title }, t.push(i));
        continue;
      }
      if (i = this.tokenizer.table(e)) {
        e = e.substring(i.raw.length), t.push(i);
        continue;
      }
      if (i = this.tokenizer.lheading(e)) {
        e = e.substring(i.raw.length), t.push(i);
        continue;
      }
      let c = e;
      if ((l = this.options.extensions) != null && l.startBlock) {
        let p = 1 / 0, f = e.slice(1), y;
        this.options.extensions.startBlock.forEach((b) => {
          y = b.call({ lexer: this }, f), typeof y == "number" && y >= 0 && (p = Math.min(p, y));
        }), p < 1 / 0 && p >= 0 && (c = e.substring(0, p + 1));
      }
      if (this.state.top && (i = this.tokenizer.paragraph(c))) {
        let p = t.at(-1);
        r && (p == null ? void 0 : p.type) === "paragraph" ? (p.raw += (p.raw.endsWith(`
`) ? "" : `
`) + i.raw, p.text += `
` + i.text, this.inlineQueue.pop(), this.inlineQueue.at(-1).src = p.text) : t.push(i), r = c.length !== e.length, e = e.substring(i.raw.length);
        continue;
      }
      if (i = this.tokenizer.text(e)) {
        e = e.substring(i.raw.length);
        let p = t.at(-1);
        (p == null ? void 0 : p.type) === "text" ? (p.raw += (p.raw.endsWith(`
`) ? "" : `
`) + i.raw, p.text += `
` + i.text, this.inlineQueue.pop(), this.inlineQueue.at(-1).src = p.text) : t.push(i);
        continue;
      }
      if (e) {
        let p = "Infinite loop on byte: " + e.charCodeAt(0);
        if (this.options.silent) {
          console.error(p);
          break;
        } else throw new Error(p);
      }
    }
    return this.state.top = !0, t;
  }
  inline(e, t = []) {
    return this.inlineQueue.push({ src: e, tokens: t }), t;
  }
  inlineTokens(e, t = []) {
    var c, p, f, y, b;
    this.tokenizer.lexer = this;
    let r = e, s = null;
    if (this.tokens.links) {
      let d = Object.keys(this.tokens.links);
      if (d.length > 0) for (; (s = this.tokenizer.rules.inline.reflinkSearch.exec(r)) !== null; ) d.includes(s[0].slice(s[0].lastIndexOf("[") + 1, -1)) && (r = r.slice(0, s.index) + "[" + "a".repeat(s[0].length - 2) + "]" + r.slice(this.tokenizer.rules.inline.reflinkSearch.lastIndex));
    }
    for (; (s = this.tokenizer.rules.inline.anyPunctuation.exec(r)) !== null; ) r = r.slice(0, s.index) + "++" + r.slice(this.tokenizer.rules.inline.anyPunctuation.lastIndex);
    let o;
    for (; (s = this.tokenizer.rules.inline.blockSkip.exec(r)) !== null; ) o = s[2] ? s[2].length : 0, r = r.slice(0, s.index + o) + "[" + "a".repeat(s[0].length - o - 2) + "]" + r.slice(this.tokenizer.rules.inline.blockSkip.lastIndex);
    r = ((p = (c = this.options.hooks) == null ? void 0 : c.emStrongMask) == null ? void 0 : p.call({ lexer: this }, r)) ?? r;
    let l = !1, i = "";
    for (; e; ) {
      l || (i = ""), l = !1;
      let d;
      if ((y = (f = this.options.extensions) == null ? void 0 : f.inline) != null && y.some((z) => (d = z.call({ lexer: this }, e, t)) ? (e = e.substring(d.raw.length), t.push(d), !0) : !1)) continue;
      if (d = this.tokenizer.escape(e)) {
        e = e.substring(d.raw.length), t.push(d);
        continue;
      }
      if (d = this.tokenizer.tag(e)) {
        e = e.substring(d.raw.length), t.push(d);
        continue;
      }
      if (d = this.tokenizer.link(e)) {
        e = e.substring(d.raw.length), t.push(d);
        continue;
      }
      if (d = this.tokenizer.reflink(e, this.tokens.links)) {
        e = e.substring(d.raw.length);
        let z = t.at(-1);
        d.type === "text" && (z == null ? void 0 : z.type) === "text" ? (z.raw += d.raw, z.text += d.text) : t.push(d);
        continue;
      }
      if (d = this.tokenizer.emStrong(e, r, i)) {
        e = e.substring(d.raw.length), t.push(d);
        continue;
      }
      if (d = this.tokenizer.codespan(e)) {
        e = e.substring(d.raw.length), t.push(d);
        continue;
      }
      if (d = this.tokenizer.br(e)) {
        e = e.substring(d.raw.length), t.push(d);
        continue;
      }
      if (d = this.tokenizer.del(e, r, i)) {
        e = e.substring(d.raw.length), t.push(d);
        continue;
      }
      if (d = this.tokenizer.autolink(e)) {
        e = e.substring(d.raw.length), t.push(d);
        continue;
      }
      if (!this.state.inLink && (d = this.tokenizer.url(e))) {
        e = e.substring(d.raw.length), t.push(d);
        continue;
      }
      let I = e;
      if ((b = this.options.extensions) != null && b.startInline) {
        let z = 1 / 0, T = e.slice(1), J;
        this.options.extensions.startInline.forEach((ne) => {
          J = ne.call({ lexer: this }, T), typeof J == "number" && J >= 0 && (z = Math.min(z, J));
        }), z < 1 / 0 && z >= 0 && (I = e.substring(0, z + 1));
      }
      if (d = this.tokenizer.inlineText(I)) {
        e = e.substring(d.raw.length), d.raw.slice(-1) !== "_" && (i = d.raw.slice(-1)), l = !0;
        let z = t.at(-1);
        (z == null ? void 0 : z.type) === "text" ? (z.raw += d.raw, z.text += d.text) : t.push(d);
        continue;
      }
      if (e) {
        let z = "Infinite loop on byte: " + e.charCodeAt(0);
        if (this.options.silent) {
          console.error(z);
          break;
        } else throw new Error(z);
      }
    }
    return t;
  }
}, Ye = class {
  constructor(n) {
    Y(this, "options");
    Y(this, "parser");
    this.options = n || ze;
  }
  space(n) {
    return "";
  }
  code({ text: n, lang: e, escaped: t }) {
    var o;
    let r = (o = (e || "").match(se.notSpaceStart)) == null ? void 0 : o[0], s = n.replace(se.endingNewline, "") + `
`;
    return r ? '<pre><code class="language-' + xe(r) + '">' + (t ? s : xe(s, !0)) + `</code></pre>
` : "<pre><code>" + (t ? s : xe(s, !0)) + `</code></pre>
`;
  }
  blockquote({ tokens: n }) {
    return `<blockquote>
${this.parser.parse(n)}</blockquote>
`;
  }
  html({ text: n }) {
    return n;
  }
  def(n) {
    return "";
  }
  heading({ tokens: n, depth: e }) {
    return `<h${e}>${this.parser.parseInline(n)}</h${e}>
`;
  }
  hr(n) {
    return `<hr>
`;
  }
  list(n) {
    let e = n.ordered, t = n.start, r = "";
    for (let l = 0; l < n.items.length; l++) {
      let i = n.items[l];
      r += this.listitem(i);
    }
    let s = e ? "ol" : "ul", o = e && t !== 1 ? ' start="' + t + '"' : "";
    return "<" + s + o + `>
` + r + "</" + s + `>
`;
  }
  listitem(n) {
    return `<li>${this.parser.parse(n.tokens)}</li>
`;
  }
  checkbox({ checked: n }) {
    return "<input " + (n ? 'checked="" ' : "") + 'disabled="" type="checkbox"> ';
  }
  paragraph({ tokens: n }) {
    return `<p>${this.parser.parseInline(n)}</p>
`;
  }
  table(n) {
    let e = "", t = "";
    for (let s = 0; s < n.header.length; s++) t += this.tablecell(n.header[s]);
    e += this.tablerow({ text: t });
    let r = "";
    for (let s = 0; s < n.rows.length; s++) {
      let o = n.rows[s];
      t = "";
      for (let l = 0; l < o.length; l++) t += this.tablecell(o[l]);
      r += this.tablerow({ text: t });
    }
    return r && (r = `<tbody>${r}</tbody>`), `<table>
<thead>
` + e + `</thead>
` + r + `</table>
`;
  }
  tablerow({ text: n }) {
    return `<tr>
${n}</tr>
`;
  }
  tablecell(n) {
    let e = this.parser.parseInline(n.tokens), t = n.header ? "th" : "td";
    return (n.align ? `<${t} align="${n.align}">` : `<${t}>`) + e + `</${t}>
`;
  }
  strong({ tokens: n }) {
    return `<strong>${this.parser.parseInline(n)}</strong>`;
  }
  em({ tokens: n }) {
    return `<em>${this.parser.parseInline(n)}</em>`;
  }
  codespan({ text: n }) {
    return `<code>${xe(n, !0)}</code>`;
  }
  br(n) {
    return "<br>";
  }
  del({ tokens: n }) {
    return `<del>${this.parser.parseInline(n)}</del>`;
  }
  link({ href: n, title: e, tokens: t }) {
    let r = this.parser.parseInline(t), s = vt(n);
    if (s === null) return r;
    n = s;
    let o = '<a href="' + n + '"';
    return e && (o += ' title="' + xe(e) + '"'), o += ">" + r + "</a>", o;
  }
  image({ href: n, title: e, text: t, tokens: r }) {
    r && (t = this.parser.parseInline(r, this.parser.textRenderer));
    let s = vt(n);
    if (s === null) return xe(t);
    n = s;
    let o = `<img src="${n}" alt="${xe(t)}"`;
    return e && (o += ` title="${xe(e)}"`), o += ">", o;
  }
  text(n) {
    return "tokens" in n && n.tokens ? this.parser.parseInline(n.tokens) : "escaped" in n && n.escaped ? n.text : xe(n.text);
  }
}, dt = class {
  strong({ text: n }) {
    return n;
  }
  em({ text: n }) {
    return n;
  }
  codespan({ text: n }) {
    return n;
  }
  del({ text: n }) {
    return n;
  }
  html({ text: n }) {
    return n;
  }
  text({ text: n }) {
    return n;
  }
  link({ text: n }) {
    return "" + n;
  }
  image({ text: n }) {
    return "" + n;
  }
  br() {
    return "";
  }
  checkbox({ raw: n }) {
    return n;
  }
}, ge = class nt {
  constructor(e) {
    Y(this, "options");
    Y(this, "renderer");
    Y(this, "textRenderer");
    this.options = e || ze, this.options.renderer = this.options.renderer || new Ye(), this.renderer = this.options.renderer, this.renderer.options = this.options, this.renderer.parser = this, this.textRenderer = new dt();
  }
  static parse(e, t) {
    return new nt(t).parse(e);
  }
  static parseInline(e, t) {
    return new nt(t).parseInline(e);
  }
  parse(e) {
    var r, s;
    this.renderer.parser = this;
    let t = "";
    for (let o = 0; o < e.length; o++) {
      let l = e[o];
      if ((s = (r = this.options.extensions) == null ? void 0 : r.renderers) != null && s[l.type]) {
        let c = l, p = this.options.extensions.renderers[c.type].call({ parser: this }, c);
        if (p !== !1 || !["space", "hr", "heading", "code", "table", "blockquote", "list", "html", "def", "paragraph", "text"].includes(c.type)) {
          t += p || "";
          continue;
        }
      }
      let i = l;
      switch (i.type) {
        case "space": {
          t += this.renderer.space(i);
          break;
        }
        case "hr": {
          t += this.renderer.hr(i);
          break;
        }
        case "heading": {
          t += this.renderer.heading(i);
          break;
        }
        case "code": {
          t += this.renderer.code(i);
          break;
        }
        case "table": {
          t += this.renderer.table(i);
          break;
        }
        case "blockquote": {
          t += this.renderer.blockquote(i);
          break;
        }
        case "list": {
          t += this.renderer.list(i);
          break;
        }
        case "checkbox": {
          t += this.renderer.checkbox(i);
          break;
        }
        case "html": {
          t += this.renderer.html(i);
          break;
        }
        case "def": {
          t += this.renderer.def(i);
          break;
        }
        case "paragraph": {
          t += this.renderer.paragraph(i);
          break;
        }
        case "text": {
          t += this.renderer.text(i);
          break;
        }
        default: {
          let c = 'Token with "' + i.type + '" type was not found.';
          if (this.options.silent) return console.error(c), "";
          throw new Error(c);
        }
      }
    }
    return t;
  }
  parseInline(e, t = this.renderer) {
    var s, o;
    this.renderer.parser = this;
    let r = "";
    for (let l = 0; l < e.length; l++) {
      let i = e[l];
      if ((o = (s = this.options.extensions) == null ? void 0 : s.renderers) != null && o[i.type]) {
        let p = this.options.extensions.renderers[i.type].call({ parser: this }, i);
        if (p !== !1 || !["escape", "html", "link", "image", "strong", "em", "codespan", "br", "del", "text"].includes(i.type)) {
          r += p || "";
          continue;
        }
      }
      let c = i;
      switch (c.type) {
        case "escape": {
          r += t.text(c);
          break;
        }
        case "html": {
          r += t.html(c);
          break;
        }
        case "link": {
          r += t.link(c);
          break;
        }
        case "image": {
          r += t.image(c);
          break;
        }
        case "checkbox": {
          r += t.checkbox(c);
          break;
        }
        case "strong": {
          r += t.strong(c);
          break;
        }
        case "em": {
          r += t.em(c);
          break;
        }
        case "codespan": {
          r += t.codespan(c);
          break;
        }
        case "br": {
          r += t.br(c);
          break;
        }
        case "del": {
          r += t.del(c);
          break;
        }
        case "text": {
          r += t.text(c);
          break;
        }
        default: {
          let p = 'Token with "' + c.type + '" type was not found.';
          if (this.options.silent) return console.error(p), "";
          throw new Error(p);
        }
      }
    }
    return r;
  }
}, We, Fe = (We = class {
  constructor(n) {
    Y(this, "options");
    Y(this, "block");
    this.options = n || ze;
  }
  preprocess(n) {
    return n;
  }
  postprocess(n) {
    return n;
  }
  processAllTokens(n) {
    return n;
  }
  emStrongMask(n) {
    return n;
  }
  provideLexer(n = this.block) {
    return n ? de.lex : de.lexInline;
  }
  provideParser(n = this.block) {
    return n ? ge.parse : ge.parseInline;
  }
}, Y(We, "passThroughHooks", /* @__PURE__ */ new Set(["preprocess", "postprocess", "processAllTokens", "emStrongMask"])), Y(We, "passThroughHooksRespectAsync", /* @__PURE__ */ new Set(["preprocess", "postprocess", "processAllTokens"])), We), Rr = class {
  constructor(...n) {
    Y(this, "defaults", it());
    Y(this, "options", this.setOptions);
    Y(this, "parse", this.parseMarkdown(!0));
    Y(this, "parseInline", this.parseMarkdown(!1));
    Y(this, "Parser", ge);
    Y(this, "Renderer", Ye);
    Y(this, "TextRenderer", dt);
    Y(this, "Lexer", de);
    Y(this, "Tokenizer", Ne);
    Y(this, "Hooks", Fe);
    this.use(...n);
  }
  walkTokens(n, e) {
    var r, s;
    let t = [];
    for (let o of n) switch (t = t.concat(e.call(this, o)), o.type) {
      case "table": {
        let l = o;
        for (let i of l.header) t = t.concat(this.walkTokens(i.tokens, e));
        for (let i of l.rows) for (let c of i) t = t.concat(this.walkTokens(c.tokens, e));
        break;
      }
      case "list": {
        let l = o;
        t = t.concat(this.walkTokens(l.items, e));
        break;
      }
      default: {
        let l = o;
        (s = (r = this.defaults.extensions) == null ? void 0 : r.childTokens) != null && s[l.type] ? this.defaults.extensions.childTokens[l.type].forEach((i) => {
          let c = l[i].flat(1 / 0);
          t = t.concat(this.walkTokens(c, e));
        }) : l.tokens && (t = t.concat(this.walkTokens(l.tokens, e)));
      }
    }
    return t;
  }
  use(...n) {
    let e = this.defaults.extensions || { renderers: {}, childTokens: {} };
    return n.forEach((t) => {
      let r = { ...t };
      if (r.async = this.defaults.async || r.async || !1, t.extensions && (t.extensions.forEach((s) => {
        if (!s.name) throw new Error("extension name required");
        if ("renderer" in s) {
          let o = e.renderers[s.name];
          o ? e.renderers[s.name] = function(...l) {
            let i = s.renderer.apply(this, l);
            return i === !1 && (i = o.apply(this, l)), i;
          } : e.renderers[s.name] = s.renderer;
        }
        if ("tokenizer" in s) {
          if (!s.level || s.level !== "block" && s.level !== "inline") throw new Error("extension level must be 'block' or 'inline'");
          let o = e[s.level];
          o ? o.unshift(s.tokenizer) : e[s.level] = [s.tokenizer], s.start && (s.level === "block" ? e.startBlock ? e.startBlock.push(s.start) : e.startBlock = [s.start] : s.level === "inline" && (e.startInline ? e.startInline.push(s.start) : e.startInline = [s.start]));
        }
        "childTokens" in s && s.childTokens && (e.childTokens[s.name] = s.childTokens);
      }), r.extensions = e), t.renderer) {
        let s = this.defaults.renderer || new Ye(this.defaults);
        for (let o in t.renderer) {
          if (!(o in s)) throw new Error(`renderer '${o}' does not exist`);
          if (["options", "parser"].includes(o)) continue;
          let l = o, i = t.renderer[l], c = s[l];
          s[l] = (...p) => {
            let f = i.apply(s, p);
            return f === !1 && (f = c.apply(s, p)), f || "";
          };
        }
        r.renderer = s;
      }
      if (t.tokenizer) {
        let s = this.defaults.tokenizer || new Ne(this.defaults);
        for (let o in t.tokenizer) {
          if (!(o in s)) throw new Error(`tokenizer '${o}' does not exist`);
          if (["options", "rules", "lexer"].includes(o)) continue;
          let l = o, i = t.tokenizer[l], c = s[l];
          s[l] = (...p) => {
            let f = i.apply(s, p);
            return f === !1 && (f = c.apply(s, p)), f;
          };
        }
        r.tokenizer = s;
      }
      if (t.hooks) {
        let s = this.defaults.hooks || new Fe();
        for (let o in t.hooks) {
          if (!(o in s)) throw new Error(`hook '${o}' does not exist`);
          if (["options", "block"].includes(o)) continue;
          let l = o, i = t.hooks[l], c = s[l];
          Fe.passThroughHooks.has(o) ? s[l] = (p) => {
            if (this.defaults.async && Fe.passThroughHooksRespectAsync.has(o)) return (async () => {
              let y = await i.call(s, p);
              return c.call(s, y);
            })();
            let f = i.call(s, p);
            return c.call(s, f);
          } : s[l] = (...p) => {
            if (this.defaults.async) return (async () => {
              let y = await i.apply(s, p);
              return y === !1 && (y = await c.apply(s, p)), y;
            })();
            let f = i.apply(s, p);
            return f === !1 && (f = c.apply(s, p)), f;
          };
        }
        r.hooks = s;
      }
      if (t.walkTokens) {
        let s = this.defaults.walkTokens, o = t.walkTokens;
        r.walkTokens = function(l) {
          let i = [];
          return i.push(o.call(this, l)), s && (i = i.concat(s.call(this, l))), i;
        };
      }
      this.defaults = { ...this.defaults, ...r };
    }), this;
  }
  setOptions(n) {
    return this.defaults = { ...this.defaults, ...n }, this;
  }
  lexer(n, e) {
    return de.lex(n, e ?? this.defaults);
  }
  parser(n, e) {
    return ge.parse(n, e ?? this.defaults);
  }
  parseMarkdown(n) {
    return (e, t) => {
      let r = { ...t }, s = { ...this.defaults, ...r }, o = this.onError(!!s.silent, !!s.async);
      if (this.defaults.async === !0 && r.async === !1) return o(new Error("marked(): The async option was set to true by an extension. Remove async: false from the parse options object to return a Promise."));
      if (typeof e > "u" || e === null) return o(new Error("marked(): input parameter is undefined or null"));
      if (typeof e != "string") return o(new Error("marked(): input parameter is of type " + Object.prototype.toString.call(e) + ", string expected"));
      if (s.hooks && (s.hooks.options = s, s.hooks.block = n), s.async) return (async () => {
        let l = s.hooks ? await s.hooks.preprocess(e) : e, i = await (s.hooks ? await s.hooks.provideLexer(n) : n ? de.lex : de.lexInline)(l, s), c = s.hooks ? await s.hooks.processAllTokens(i) : i;
        s.walkTokens && await Promise.all(this.walkTokens(c, s.walkTokens));
        let p = await (s.hooks ? await s.hooks.provideParser(n) : n ? ge.parse : ge.parseInline)(c, s);
        return s.hooks ? await s.hooks.postprocess(p) : p;
      })().catch(o);
      try {
        s.hooks && (e = s.hooks.preprocess(e));
        let l = (s.hooks ? s.hooks.provideLexer(n) : n ? de.lex : de.lexInline)(e, s);
        s.hooks && (l = s.hooks.processAllTokens(l)), s.walkTokens && this.walkTokens(l, s.walkTokens);
        let i = (s.hooks ? s.hooks.provideParser(n) : n ? ge.parse : ge.parseInline)(l, s);
        return s.hooks && (i = s.hooks.postprocess(i)), i;
      } catch (l) {
        return o(l);
      }
    };
  }
  onError(n, e) {
    return (t) => {
      if (t.message += `
Please report this to https://github.com/markedjs/marked.`, n) {
        let r = "<p>An error occurred:</p><pre>" + xe(t.message + "", !0) + "</pre>";
        return e ? Promise.resolve(r) : r;
      }
      if (e) return Promise.reject(t);
      throw t;
    };
  }
}, Re = new Rr();
function N(n, e) {
  return Re.parse(n, e);
}
N.options = N.setOptions = function(n) {
  return Re.setOptions(n), N.defaults = Re.defaults, Ft(N.defaults), N;
};
N.getDefaults = it;
N.defaults = ze;
N.use = function(...n) {
  return Re.use(...n), N.defaults = Re.defaults, Ft(N.defaults), N;
};
N.walkTokens = function(n, e) {
  return Re.walkTokens(n, e);
};
N.parseInline = Re.parseInline;
N.Parser = ge;
N.parser = ge.parse;
N.Renderer = Ye;
N.TextRenderer = dt;
N.Lexer = de;
N.lexer = de.lex;
N.Tokenizer = Ne;
N.Hooks = Fe;
N.parse = N;
N.options;
N.setOptions;
N.use;
N.walkTokens;
N.parseInline;
ge.parse;
de.lex;
function Oe(n) {
  return n.replace(/\s+/g, " ");
}
function Ae(n, e = {}) {
  const t = [];
  for (const r of n)
    switch (r.type) {
      case "text":
        t.push({ text: Oe(r.text), ...e });
        break;
      case "strong":
        t.push(...Ae(r.tokens || [], { ...e, bold: !0 }));
        break;
      case "em":
        t.push(...Ae(r.tokens || [], { ...e, italic: !0 }));
        break;
      case "codespan":
        t.push({ text: Oe(r.text), ...e, code: !0 });
        break;
      case "link":
        t.push(...Ae(r.tokens || [], { ...e, link: r.href }));
        break;
      case "del":
        t.push(...Ae(r.tokens || [], { ...e, strikethrough: !0 }));
        break;
      case "br":
        t.push({ text: `
`, ...e });
        break;
      case "escape":
        t.push({ text: Oe(r.text), ...e });
        break;
      default:
        "text" in r && t.push({ text: Oe(r.text), ...e });
    }
  return t;
}
function At(n) {
  return n.map((e) => e.text).join("");
}
function rt(n) {
  const e = [];
  for (const t of n)
    switch (t.type) {
      case "heading": {
        const r = Ae(t.tokens || []);
        e.push({
          type: "heading",
          text: At(r),
          segments: r,
          tag: `h${t.depth || 2}`
        });
        break;
      }
      case "paragraph": {
        const r = Ae(t.tokens || []);
        e.push({
          type: "paragraph",
          text: At(r),
          segments: r
        });
        break;
      }
      case "list": {
        const r = t, s = r.items.map((o) => {
          const l = rt(o.tokens || []);
          if (l.length === 1)
            return { ...l[0], type: "paragraph" };
          const i = l.flatMap((c) => c.segments || [{ text: c.text }]);
          return {
            type: "paragraph",
            text: l.map((c) => c.text).join(" "),
            segments: i
          };
        });
        e.push({
          type: "list",
          text: s.map((o) => o.text).join(`
`),
          tag: r.ordered ? "ol" : "ul",
          items: s
        });
        break;
      }
      case "blockquote": {
        const r = rt(t.tokens || []), s = r.flatMap((o) => o.segments || [{ text: o.text }]);
        e.push({
          type: "blockquote",
          text: r.map((o) => o.text).join(" "),
          segments: s
        });
        break;
      }
      case "code": {
        e.push({
          type: "code",
          text: t.text,
          language: t.lang || void 0
        });
        break;
      }
      case "hr": {
        e.push({ type: "hr", text: "" });
        break;
      }
      case "space":
        break;
      default:
        "text" in t && t.text && e.push({
          type: "paragraph",
          text: t.text
        });
    }
  return e;
}
function $r(n) {
  const e = N.lexer(n);
  return rt(e);
}
function zr(n) {
  return n.map((e) => {
    switch (e.type) {
      case "heading":
        return `${e.tag === "h1" ? "#" : e.tag === "h3" ? "###" : "##"} ${qe(e.segments) || e.text}`;
      case "paragraph":
        return qe(e.segments) || e.text;
      case "list": {
        const t = e.tag === "ol" ? (r) => `${r + 1}. ` : () => "- ";
        return (e.items || []).map(
          (r, s) => `${t(s)}${qe(r.segments) || r.text}`
        ).join(`
`);
      }
      case "blockquote":
        return `> ${qe(e.segments) || e.text}`;
      case "code":
        return `\`\`\`${e.language || ""}
${e.text}
\`\`\``;
      case "hr":
        return "---";
      default:
        return e.text;
    }
  }).join(`

`);
}
function qe(n) {
  return n ? n.map((e) => {
    let t = e.text;
    return e.code && (t = `\`${t}\``), e.bold && e.italic ? t = `***${t}***` : e.bold ? t = `**${t}**` : e.italic && (t = `*${t}*`), e.strikethrough && (t = `~~${t}~~`), e.link && (t = `[${t}](${e.link})`), t;
  }).join("") : "";
}
const Mt = {
  bodyFont: "16px Lato, sans-serif"
}, Ar = st(function({
  blocks: e,
  layout: t,
  onLayoutChange: r,
  onBlocksChange: s,
  resolveImageUrl: o,
  config: l,
  availableFonts: i,
  availableInitialFonts: c,
  height: p,
  expandable: f,
  width: y
}, b) {
  var xt, kt, mt, bt, yt, wt, St;
  const d = ue(null), I = ue(null), z = ue(null), T = ue(null), [J, ne] = ke(() => zr(e)), [B, E] = ke(null), [$, D] = ke(null), [h, w] = ke(!1), [k, H] = ke(-1), [Q, W] = ke(/* @__PURE__ */ new Set([(l == null ? void 0 : l.defaultMode) || "layout"])), A = (a, u) => {
    W(u ? (x) => {
      const L = new Set(x);
      return L.has(a) ? L.size > 1 && L.delete(a) : L.add(a), L;
    } : /* @__PURE__ */ new Set([a]));
  }, O = ue(null), m = ue(null), S = ue(null), P = ue(!1), g = t || { images: [] }, X = ae((a, u, x, L, _) => {
    const q = i == null ? void 0 : i.find((Le) => Le.name === a), oe = q ? q.bodyFont.replace(/^\d+px\s*/, "") : Mt.bodyFont.replace(/^\d+px\s*/, ""), re = u || 16, R = `${re}px ${oe}`, G = (q == null ? void 0 : q.bodyLineHeight) ?? Math.round(re * 1.6), U = (q == null ? void 0 : q.headingLineHeight) ?? Math.round(re * 1.5 * 1.4), pe = c == null ? void 0 : c.find((Le) => Le.name === L), le = (pe == null ? void 0 : pe.fontFamily) || "serif", Ie = x ? `${_ || 96}px ${le}` : void 0;
    return { bodyFontCSS: R, bodyLineHeight: G, headingLineHeight: U, initialCapFontCSS: Ie };
  }, [i, c]), C = ae((a) => {
    var L;
    const u = X(a.fontFamily, a.fontSize, a.initialCap, a.initialCapFont, a.initialCapSize), x = (L = a.breakpoints) == null ? void 0 : L.map((_) => {
      const q = X(
        _.fontFamily ?? a.fontFamily,
        _.fontSize ?? a.fontSize,
        _.initialCap ?? a.initialCap,
        _.initialCapFont ?? a.initialCapFont,
        _.initialCapSize ?? a.initialCapSize
      );
      return { ..._, ...q };
    });
    r({ ...a, ...u, breakpoints: x });
  }, [r, X]), V = i == null ? void 0 : i.find((a) => a.name === g.fontFamily);
  V ? V.bodyFont.replace(/^\d+px\s*/, "") : Mt.bodyFont.replace(/^\d+px\s*/, ""), g.fontSize, {
    ...V != null && V.bodyLineHeight ? { bodyLineHeight: V.bodyLineHeight } : {},
    ...V != null && V.headingLineHeight ? { headingLineHeight: V.headingLineHeight } : {}
  }, { ...l };
  const M = k === -1 ? null : (xt = g.breakpoints) == null ? void 0 : xt[k], j = (M == null ? void 0 : M.images) || g.images, ie = (M == null ? void 0 : M.columns) ?? g.columns, ee = (M == null ? void 0 : M.fontFamily) ?? g.fontFamily, te = (M == null ? void 0 : M.fontSize) ?? g.fontSize, Z = (M == null ? void 0 : M.initialCap) ?? g.initialCap ?? !1, ce = (M == null ? void 0 : M.initialCapFont) ?? g.initialCapFont, me = (M == null ? void 0 : M.initialCapSize) ?? g.initialCapSize, we = (M == null ? void 0 : M.initialCapOffsetX) ?? g.initialCapOffsetX ?? 0, fe = (M == null ? void 0 : M.initialCapOffsetY) ?? g.initialCapOffsetY ?? 0, be = (a, u) => {
    if (k === -1)
      C({ ...g, images: a, ...u });
    else {
      const x = [...g.breakpoints || []];
      x[k] = { ...x[k], images: a }, C({ ...g, breakpoints: x });
    }
  }, Ge = (a) => {
    if (k === -1)
      C({ ...g, columns: a });
    else {
      const u = [...g.breakpoints || []];
      u[k] = { ...u[k], columns: a }, C({ ...g, breakpoints: u });
    }
  };
  an(b, () => ({
    addImage: (a) => {
      var L;
      const u = k === -1 ? null : (L = g.breakpoints) == null ? void 0 : L[k], x = (u == null ? void 0 : u.images) || g.images;
      if (k === -1)
        C({ ...g, images: [...x, a] });
      else {
        const _ = [...g.breakpoints || []];
        _[k] = { ..._[k], images: [...x, a] }, C({ ...g, breakpoints: _ });
      }
    }
  }), [g, k, C]);
  const ve = (a, u) => {
    const x = {
      maxWidth: a,
      name: u,
      images: j.map((_) => ({ ..._ })),
      columns: ie,
      editorWidth: a,
      fontFamily: ee,
      fontSize: te,
      initialCap: Z,
      initialCapFont: ce,
      initialCapSize: me,
      initialCapOffsetX: we,
      initialCapOffsetY: fe
    }, L = [...g.breakpoints || [], x];
    C({ ...g, breakpoints: L }), H(L.length - 1);
  }, Yt = (a) => {
    const u = (g.breakpoints || []).filter((x, L) => L !== a);
    C({ ...g, breakpoints: u }), k === a ? H(-1) : k > a && H(k - 1);
  }, gt = ue(!0);
  Me(() => {
    if (gt.current) {
      gt.current = !1;
      return;
    }
  }, [e]);
  const jt = ae((a) => {
    const u = a.target.value;
    ne(u), s && s($r(u));
  }, [s]), Zt = ae(() => {
    const a = T.current;
    a && (a.style.height = "auto", a.style.height = a.scrollHeight + "px");
  }, []);
  Me(() => {
    Zt();
  }, [J, Q]);
  const Se = ae((a, u, x, L, _) => {
    const q = T.current;
    q && (q.focus(), q.selectionStart = a, q.selectionEnd = u, document.execCommand("insertText", !1, x), L !== void 0 && (q.selectionStart = L, q.selectionEnd = _ ?? L));
  }, []), Gt = ae((a) => {
    const u = a.currentTarget, { selectionStart: x, selectionEnd: L, value: _ } = u, q = x !== L, oe = _.slice(x, L), re = (R, G) => {
      a.preventDefault();
      const U = R + oe + G;
      Se(x, L, U, x + R.length, L + R.length);
    };
    if ((a.metaKey || a.ctrlKey) && a.key === "b") {
      re("**", "**");
      return;
    }
    if ((a.metaKey || a.ctrlKey) && a.key === "i") {
      re("*", "*");
      return;
    }
    if ((a.metaKey || a.ctrlKey) && a.key === "k") {
      a.preventDefault();
      const R = q ? oe : "link text", G = `[${R}](url)`, U = x + R.length + 3;
      Se(x, L, G, U, U + 3);
      return;
    }
    if (q && ["*", "`", "~"].includes(a.key)) {
      a.preventDefault(), a.key === "~" ? re("~~", "~~") : re(a.key, a.key);
      return;
    }
    if (a.key === "Tab") {
      a.preventDefault();
      const R = _.lastIndexOf(`
`, x - 1) + 1, G = _.indexOf(`
`, x);
      G === -1 && _.length, a.shiftKey ? _.slice(R, R + 2) === "  " && Se(R, R + 2, "", Math.max(R, x - 2)) : Se(R, R, "  ", x + 2);
      return;
    }
    if (a.key === "Enter") {
      const R = _.lastIndexOf(`
`, x - 1) + 1, U = _.slice(R, x).match(/^(\s*)([-*]|\d+\.)\s(.*)/);
      if (U) {
        const [, pe, le, he] = U;
        if (he.trim() === "") {
          a.preventDefault(), Se(R, x, `
`, R + 1);
          return;
        }
        a.preventDefault();
        const Ie = le.match(/^\d+\./) ? `${parseInt(le) + 1}.` : le, Le = `
${pe}${Ie} `;
        Se(x, L, Le, x + Le.length);
        return;
      }
    }
  }, [Se]), Qe = (a) => {
    var u;
    return (u = z.current) == null ? void 0 : u.querySelector(`img[data-image-index="${a}"]`);
  }, Qt = ae((a) => {
    if ($ !== null) {
      const u = Qe($);
      if (u) {
        const x = u.getBoundingClientRect(), L = (a.clientX - x.left) / x.width, _ = (a.clientY - x.top) / x.height, q = j[$].polygon || [], oe = 0.04;
        let re = -1, R = 1 / 0;
        for (let G = 0; G < q.length; G++) {
          const U = L - q[G].x, pe = _ - q[G].y, le = Math.sqrt(U * U + pe * pe);
          le < oe && le < R && (R = le, re = G);
        }
        re >= 0 && (S.current = { imageIndex: $, pointIndex: re, imgRect: x, startX: a.clientX, startY: a.clientY, active: !1 }, P.current = !0, a.preventDefault());
      }
    }
  }, [g, $]), Ut = ae((a, u) => {
    $ === null && (O.current = {
      imageIndex: u,
      startX: a.clientX,
      startY: a.clientY,
      origX: j[u].x,
      origImgY: j[u].y,
      active: !1
    }, E(u), a.preventDefault(), a.stopPropagation());
  }, [j, $]), Kt = ae((a, u) => {
    m.current = { imageIndex: u, startX: a.clientX, origWidth: j[u].width }, a.preventDefault(), a.stopPropagation();
  }, [j]), Vt = ae((a) => {
    var q, oe, re;
    const u = ((q = z.current) == null ? void 0 : q.offsetWidth) || 700, x = (k === -1 ? g.editorWidth : (re = (oe = g.breakpoints) == null ? void 0 : oe[k]) == null ? void 0 : re.editorWidth) || u, L = u / x, _ = L < 1 ? Math.sqrt(L) : L;
    if (O.current) {
      const R = O.current, G = a.clientX - R.startX, U = a.clientY - R.startY;
      if (!R.active && Math.abs(G) + Math.abs(U) < 3) return;
      R.active = !0;
      const pe = R.origX + G / L, le = Math.max(0, R.origImgY + U / L), he = [...j];
      he[R.imageIndex] = {
        ...he[R.imageIndex],
        x: pe,
        y: le
      }, be(he);
    }
    if (m.current) {
      const R = m.current, G = a.clientX - R.startX, U = [...j];
      U[R.imageIndex] = { ...U[R.imageIndex], width: Math.max(50, R.origWidth + G / _) }, be(U);
    }
    if (S.current) {
      const R = S.current, G = a.clientX - R.startX, U = a.clientY - R.startY;
      if (!R.active && Math.abs(G) + Math.abs(U) < 5) return;
      R.active = !0;
      const pe = Math.max(0, Math.min(1, (a.clientX - R.imgRect.left) / R.imgRect.width)), le = Math.max(0, Math.min(1, (a.clientY - R.imgRect.top) / R.imgRect.height)), he = [...j], Ie = [...he[R.imageIndex].polygon || []];
      Ie[R.pointIndex] = { x: pe, y: le }, he[R.imageIndex] = { ...he[R.imageIndex], polygon: Ie }, be(he);
    }
  }, [g, C, j, k]), ft = ae(() => {
    O.current && (O.current = null), m.current && (m.current = null), S.current && (S.current = null);
  }, []), Jt = ae((a) => {
    const u = a.target;
    if ($ !== null) {
      if (P.current) {
        P.current = !1;
        return;
      }
      const x = Qe($);
      if (!x) return;
      const L = x.getBoundingClientRect(), _ = (a.clientX - L.left) / L.width, q = (a.clientY - L.top) / L.height;
      if (_ >= -0.1 && _ <= 1.1 && q >= -0.1 && q <= 1.1) {
        const oe = [...j], re = oe[$].polygon || [];
        oe[$] = {
          ...oe[$],
          polygon: [...re, { x: Math.max(0, Math.min(1, _)), y: Math.max(0, Math.min(1, q)) }]
        }, be(oe);
      }
      return;
    }
    u.dataset.imageIndex !== void 0 ? E(parseInt(u.dataset.imageIndex)) : E(null);
  }, [$, g, C]), en = (a) => {
    const u = j.filter((x, L) => L !== a);
    be(u), E(null);
  }, Ue = (() => {
    if (B === null || !I.current) return null;
    const a = Qe(B);
    if (!a) return null;
    const u = a.getBoundingClientRect(), x = I.current.getBoundingClientRect();
    return {
      top: u.top - x.top + I.current.scrollTop - 37,
      left: u.left - x.left + I.current.scrollLeft + u.width / 2
    };
  })(), _e = B !== null ? j[B] : null, tn = [
    { key: "write", label: "Write", icon: /* @__PURE__ */ v(Rn, { size: 14 }) },
    { key: "layout", label: "Layout", icon: /* @__PURE__ */ v(Sn, { size: 14 }) }
  ];
  return /* @__PURE__ */ K("div", { style: {
    marginBottom: 20,
    width: f ? `calc(${typeof y == "number" ? y + "px" : y || "100%"} * ${Q.size})` : y,
    transition: "width 0.2s ease"
  }, children: [
    /* @__PURE__ */ K("div", { style: {
      display: "flex",
      alignItems: "center",
      background: "#ede7f6",
      borderRadius: "6px 6px 0 0",
      padding: "4px 6px",
      gap: 2,
      borderBottom: "2px solid #502581"
    }, children: [
      tn.map((a) => /* @__PURE__ */ K(
        "button",
        {
          type: "button",
          onClick: (u) => A(a.key, u.shiftKey),
          style: {
            display: "flex",
            alignItems: "center",
            gap: 5,
            padding: "6px 14px",
            fontSize: 13,
            cursor: "pointer",
            background: Q.has(a.key) ? "#502581" : "transparent",
            color: Q.has(a.key) ? "white" : "#6a4c93",
            border: "none",
            borderRadius: 4,
            fontWeight: Q.has(a.key) ? 600 : 400,
            transition: "background 0.15s, color 0.15s"
          },
          children: [
            a.icon,
            a.label
          ]
        },
        a.key
      )),
      /* @__PURE__ */ K("div", { style: { marginLeft: "auto", display: "flex", alignItems: "center", gap: 8 }, children: [
        i && i.length > 0 && /* @__PURE__ */ K(
          "select",
          {
            value: ee || "",
            onChange: (a) => {
              const u = a.target.value || void 0;
              if (k === -1)
                C({ ...g, fontFamily: u });
              else {
                const x = [...g.breakpoints || []];
                x[k] = { ...x[k], fontFamily: u }, C({ ...g, breakpoints: x });
              }
            },
            style: {
              padding: "3px 6px",
              fontSize: 12,
              cursor: "pointer",
              background: "transparent",
              color: "#6a4c93",
              border: "1px solid #d4c5e8",
              borderRadius: 3,
              outline: "none"
            },
            children: [
              /* @__PURE__ */ v("option", { value: "", children: "Default" }),
              i.map((a) => /* @__PURE__ */ v("option", { value: a.name, children: a.name }, a.name))
            ]
          }
        ),
        /* @__PURE__ */ v(
          "select",
          {
            value: te || 16,
            onChange: (a) => {
              const u = parseInt(a.target.value);
              if (k === -1)
                C({ ...g, fontSize: u });
              else {
                const x = [...g.breakpoints || []];
                x[k] = { ...x[k], fontSize: u }, C({ ...g, breakpoints: x });
              }
            },
            style: {
              padding: "3px 6px",
              fontSize: 12,
              cursor: "pointer",
              background: "transparent",
              color: "#6a4c93",
              border: "1px solid #d4c5e8",
              borderRadius: 3,
              outline: "none",
              width: 52
            },
            children: [12, 14, 16, 18, 20, 22, 24].map((a) => /* @__PURE__ */ K("option", { value: a, children: [
              a,
              "px"
            ] }, a))
          }
        ),
        /* @__PURE__ */ K("div", { style: { display: "flex", alignItems: "center", gap: 2 }, children: [
          /* @__PURE__ */ v(
            "button",
            {
              type: "button",
              title: "Initial cap",
              onClick: () => {
                const a = !Z;
                if (k === -1)
                  C({ ...g, initialCap: a });
                else {
                  const u = [...g.breakpoints || []];
                  u[k] = { ...u[k], initialCap: a }, C({ ...g, breakpoints: u });
                }
              },
              style: {
                padding: "3px 6px",
                fontSize: 12,
                cursor: "pointer",
                background: Z ? "#502581" : "transparent",
                color: Z ? "white" : "#6a4c93",
                border: "1px solid #d4c5e8",
                borderRadius: 3,
                fontWeight: Z ? 600 : 400,
                display: "flex",
                alignItems: "center",
                gap: 3
              },
              children: /* @__PURE__ */ v(xn, { size: 14 })
            }
          ),
          Z && c && c.length > 0 && /* @__PURE__ */ K(
            "select",
            {
              value: ce || "",
              onChange: (a) => {
                const u = a.target.value || void 0;
                if (k === -1)
                  C({ ...g, initialCapFont: u });
                else {
                  const x = [...g.breakpoints || []];
                  x[k] = { ...x[k], initialCapFont: u }, C({ ...g, breakpoints: x });
                }
              },
              style: {
                padding: "3px 4px",
                fontSize: 11,
                cursor: "pointer",
                background: "transparent",
                color: "#6a4c93",
                border: "1px solid #d4c5e8",
                borderRadius: 3,
                outline: "none"
              },
              children: [
                /* @__PURE__ */ v("option", { value: "", children: "Default" }),
                c.map((a) => /* @__PURE__ */ v("option", { value: a.name, children: a.name }, a.name))
              ]
            }
          ),
          Z && /* @__PURE__ */ v(
            "select",
            {
              value: me || 96,
              onChange: (a) => {
                const u = parseInt(a.target.value);
                if (k === -1)
                  C({ ...g, initialCapSize: u });
                else {
                  const x = [...g.breakpoints || []];
                  x[k] = { ...x[k], initialCapSize: u }, C({ ...g, breakpoints: x });
                }
              },
              style: {
                padding: "3px 4px",
                fontSize: 11,
                cursor: "pointer",
                background: "transparent",
                color: "#6a4c93",
                border: "1px solid #d4c5e8",
                borderRadius: 3,
                outline: "none",
                width: 52
              },
              children: [48, 64, 80, 96, 112, 128].map((a) => /* @__PURE__ */ K("option", { value: a, children: [
                a,
                "px"
              ] }, a))
            }
          ),
          Z && /* @__PURE__ */ K(sn, { children: [
            /* @__PURE__ */ v("label", { style: { fontSize: 10, color: "#6a4c93" }, children: "x" }),
            /* @__PURE__ */ v(
              "input",
              {
                type: "number",
                value: we,
                onChange: (a) => {
                  const u = parseInt(a.target.value) || 0;
                  if (k === -1)
                    C({ ...g, initialCapOffsetX: u });
                  else {
                    const x = [...g.breakpoints || []];
                    x[k] = { ...x[k], initialCapOffsetX: u }, C({ ...g, breakpoints: x });
                  }
                },
                style: { width: 44, padding: "2px 4px", fontSize: 11, border: "1px solid #d4c5e8", borderRadius: 3, color: "#6a4c93", background: "transparent", outline: "none", textAlign: "center" }
              }
            ),
            /* @__PURE__ */ v("label", { style: { fontSize: 10, color: "#6a4c93" }, children: "y" }),
            /* @__PURE__ */ v(
              "input",
              {
                type: "number",
                value: fe,
                onChange: (a) => {
                  const u = parseInt(a.target.value) || 0;
                  if (k === -1)
                    C({ ...g, initialCapOffsetY: u });
                  else {
                    const x = [...g.breakpoints || []];
                    x[k] = { ...x[k], initialCapOffsetY: u }, C({ ...g, breakpoints: x });
                  }
                },
                style: { width: 44, padding: "2px 4px", fontSize: 11, border: "1px solid #d4c5e8", borderRadius: 3, color: "#6a4c93", background: "transparent", outline: "none", textAlign: "center" }
              }
            )
          ] })
        ] }),
        /* @__PURE__ */ K("div", { style: { display: "flex", alignItems: "center", gap: 4 }, children: [
          /* @__PURE__ */ v(mn, { size: 14, color: "#6a4c93" }),
          [1, 2, 3].map((a) => /* @__PURE__ */ v(
            "button",
            {
              type: "button",
              onClick: () => Ge(a),
              style: {
                padding: "3px 8px",
                fontSize: 12,
                cursor: "pointer",
                background: (ie || 1) === a ? "#502581" : "transparent",
                color: (ie || 1) === a ? "white" : "#6a4c93",
                border: "none",
                borderRadius: 3,
                fontWeight: (ie || 1) === a ? 600 : 400
              },
              children: a
            },
            a
          ))
        ] }),
        /* @__PURE__ */ K(
          "button",
          {
            type: "button",
            title: "Preview mode (hide editor chrome)",
            onClick: () => w((a) => !a),
            style: {
              display: "flex",
              alignItems: "center",
              gap: 4,
              padding: "4px 10px",
              fontSize: 12,
              cursor: "pointer",
              background: h ? "#502581" : "transparent",
              color: h ? "white" : "#6a4c93",
              border: "none",
              borderRadius: 3,
              fontWeight: h ? 600 : 400
            },
            children: [
              /* @__PURE__ */ v(yn, { size: 14 }),
              "Preview"
            ]
          }
        )
      ] })
    ] }),
    /* @__PURE__ */ K(
      "div",
      {
        ref: d,
        style: {
          display: "flex",
          background: "white",
          border: "1px solid #ddd",
          borderRadius: "0 0 6px 6px",
          borderTop: "none",
          minHeight: p || 300
        },
        children: [
          Q.has("write") && /* @__PURE__ */ v("div", { style: { flex: 1, overflow: "auto", borderRight: Q.size > 1 ? "1px solid #ddd" : "none" }, children: /* @__PURE__ */ v(
            "textarea",
            {
              ref: T,
              autoFocus: !0,
              value: J,
              onChange: jt,
              onKeyDown: Gt,
              placeholder: `Write your content here...

Use ## for headings

Separate paragraphs with blank lines`,
              spellCheck: !1,
              style: {
                width: "100%",
                minHeight: 200,
                padding: "24px 32px",
                border: "none",
                outline: "none",
                fontFamily: "'JetBrains Mono', 'SF Mono', 'Fira Code', Consolas, monospace",
                fontSize: 18,
                lineHeight: "1.75",
                letterSpacing: "0.01em",
                color: "#2c2c2c",
                resize: "none",
                caretColor: "#502581",
                boxSizing: "border-box",
                overflow: "hidden",
                tabSize: 2
              }
            }
          ) }),
          Q.has("layout") && /* @__PURE__ */ K("div", { ref: I, style: { flex: 1, overflow: "auto", position: "relative" }, children: [
            /* @__PURE__ */ K("div", { style: {
              display: "flex",
              alignItems: "center",
              gap: 4,
              padding: "6px 10px",
              background: "#f5f0fa",
              borderBottom: "1px solid #ddd",
              fontSize: 11
            }, children: [
              /* @__PURE__ */ v(
                "button",
                {
                  type: "button",
                  onClick: () => H(-1),
                  style: {
                    padding: "3px 10px",
                    cursor: "pointer",
                    fontSize: 11,
                    background: k === -1 ? "#502581" : "transparent",
                    color: k === -1 ? "white" : "#6a4c93",
                    border: "1px solid #d4c5e8",
                    borderRadius: 3,
                    fontWeight: k === -1 ? 600 : 400
                  },
                  children: "Default"
                }
              ),
              (g.breakpoints || []).map((a, u) => /* @__PURE__ */ K("div", { style: { display: "flex", alignItems: "center" }, children: [
                /* @__PURE__ */ v(
                  "button",
                  {
                    type: "button",
                    onClick: () => H(u),
                    style: {
                      padding: "3px 10px",
                      cursor: "pointer",
                      fontSize: 11,
                      background: k === u ? "#502581" : "transparent",
                      color: k === u ? "white" : "#6a4c93",
                      border: "1px solid #d4c5e8",
                      borderRadius: "3px 0 0 3px",
                      borderRight: "none",
                      fontWeight: k === u ? 600 : 400
                    },
                    children: a.name || `≤${a.maxWidth}px`
                  }
                ),
                /* @__PURE__ */ v(
                  "button",
                  {
                    type: "button",
                    title: "Remove breakpoint",
                    onClick: (x) => {
                      x.stopPropagation(), Yt(u);
                    },
                    style: {
                      padding: "3px 5px",
                      cursor: "pointer",
                      fontSize: 11,
                      background: k === u ? "#502581" : "transparent",
                      color: k === u ? "white" : "#6a4c93",
                      border: "1px solid #d4c5e8",
                      borderRadius: "0 3px 3px 0",
                      display: "flex",
                      alignItems: "center"
                    },
                    children: /* @__PURE__ */ v(In, { size: 10 })
                  }
                )
              ] }, u)),
              /* @__PURE__ */ v(
                "button",
                {
                  type: "button",
                  title: "Add breakpoint",
                  onClick: () => {
                    const a = window.prompt("Add breakpoint — enter max width in px (e.g. 375 for mobile, 768 for tablet):", "768");
                    if (!a) return;
                    const u = parseInt(a);
                    if (isNaN(u) || u < 200) return;
                    const x = u <= 480 ? "Mobile" : u <= 1024 ? "Tablet" : `≤${u}px`;
                    ve(u, x);
                  },
                  style: {
                    padding: "3px 6px",
                    cursor: "pointer",
                    background: "transparent",
                    color: "#6a4c93",
                    border: "1px dashed #6a4c93",
                    borderRadius: 3,
                    display: "flex",
                    alignItems: "center"
                  },
                  children: /* @__PURE__ */ v(zn, { size: 12 })
                }
              )
            ] }),
            /* @__PURE__ */ v("div", { style: {
              padding: 20,
              display: "flex",
              justifyContent: "center"
            }, children: /* @__PURE__ */ v("div", { style: {
              width: k === -1 ? "100%" : ((mt = (kt = g.breakpoints) == null ? void 0 : kt[k]) == null ? void 0 : mt.maxWidth) || "100%",
              maxWidth: "100%"
            }, children: /* @__PURE__ */ v(
              Et,
              {
                containerRef: z,
                blocks: e,
                layout: k === -1 ? { ...g, breakpoints: void 0 } : { images: j, columns: ie, editorWidth: (yt = (bt = g.breakpoints) == null ? void 0 : bt[k]) == null ? void 0 : yt.editorWidth, fontFamily: ee, fontSize: te, initialCap: Z, initialCapFont: ce, initialCapSize: me, initialCapOffsetX: we, initialCapOffsetY: fe },
                config: l,
                availableFonts: i,
                availableInitialFonts: c,
                resolveImageUrl: o,
                editorMode: h ? void 0 : {
                  selectedImageIndex: B,
                  drawingPolygonIndex: $,
                  onImageMouseDown: Ut,
                  onResizeMouseDown: Kt,
                  onBackgroundMouseDown: Qt,
                  onBackgroundClick: Jt,
                  onMouseMove: Vt,
                  onMouseUp: ft,
                  onMouseLeave: ft
                }
              }
            ) }) }),
            !h && B !== null && Ue && _e && /* @__PURE__ */ v("div", { style: {
              position: "absolute",
              top: Ue.top,
              left: Ue.left,
              transform: "translateX(-50%)",
              display: "flex",
              gap: 2,
              padding: "4px 6px",
              background: "#333",
              borderRadius: 4,
              zIndex: 150,
              boxShadow: "0 2px 8px rgba(0,0,0,0.3)"
            }, children: [
              { label: (wt = _e.polygon) != null && wt.length ? `◇${_e.polygon.length}` : "◇", title: "Polygon", active: $ === B, fn: () => {
                D($ === B ? null : B);
              } },
              ...$ === B && ((St = _e.polygon) != null && St.length) ? [{
                label: "Reset",
                title: "Reset polygon",
                fn: () => {
                  const a = [...j];
                  a[B] = { ...a[B], polygon: [] }, be(a);
                }
              }] : [],
              { label: "×", title: "Remove", fn: () => en(B) }
            ].map((a, u) => /* @__PURE__ */ v(
              "button",
              {
                type: "button",
                title: a.title,
                onClick: (x) => {
                  x.stopPropagation(), a.fn();
                },
                style: {
                  padding: "3px 8px",
                  border: "none",
                  cursor: "pointer",
                  fontSize: 13,
                  borderRadius: 2,
                  background: a.active ? "#e67e22" : "transparent",
                  color: "white",
                  fontWeight: a.active ? 600 : 400,
                  boxShadow: a.active ? "0 0 6px rgba(230,126,34,0.5)" : "none"
                },
                onMouseEnter: (x) => {
                  a.active || (x.target.style.background = "rgba(255,255,255,0.2)");
                },
                onMouseLeave: (x) => {
                  a.active || (x.target.style.background = "transparent");
                },
                children: a.label
              },
              u
            )) })
          ] })
        ]
      }
    )
  ] });
});
function Mr(n) {
  return /* @__PURE__ */ v(Et, { ...n });
}
export {
  Ar as Editor,
  Mr as Renderer,
  zr as blocksToMarkdown,
  $r as parseMarkdown
};
