// react-pretext-editor/dist/layout-BDxeR5zj.js
const ve = [
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "S",
  "B",
  "S",
  "WS",
  "B",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "B",
  "B",
  "B",
  "S",
  "WS",
  "ON",
  "ON",
  "ET",
  "ET",
  "ET",
  "ON",
  "ON",
  "ON",
  "ON",
  "ON",
  "ON",
  "CS",
  "ON",
  "CS",
  "ON",
  "EN",
  "EN",
  "EN",
  "EN",
  "EN",
  "EN",
  "EN",
  "EN",
  "EN",
  "EN",
  "ON",
  "ON",
  "ON",
  "ON",
  "ON",
  "ON",
  "ON",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "ON",
  "ON",
  "ON",
  "ON",
  "ON",
  "ON",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "ON",
  "ON",
  "ON",
  "ON",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "B",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "BN",
  "CS",
  "ON",
  "ET",
  "ET",
  "ET",
  "ET",
  "ON",
  "ON",
  "ON",
  "ON",
  "L",
  "ON",
  "ON",
  "ON",
  "ON",
  "ON",
  "ET",
  "ET",
  "EN",
  "EN",
  "ON",
  "L",
  "ON",
  "ON",
  "ON",
  "EN",
  "L",
  "ON",
  "ON",
  "ON",
  "ON",
  "ON",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "ON",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "ON",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L",
  "L"
], Ie = [
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "CS",
  "AL",
  "ON",
  "ON",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AN",
  "AN",
  "AN",
  "AN",
  "AN",
  "AN",
  "AN",
  "AN",
  "AN",
  "AN",
  "ET",
  "AN",
  "AN",
  "AL",
  "AL",
  "AL",
  "NSM",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "ON",
  "NSM",
  "NSM",
  "NSM",
  "NSM",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL",
  "AL"
];
function Re(e) {
  return e <= 255 ? ve[e] : 1424 <= e && e <= 1524 ? "R" : 1536 <= e && e <= 1791 ? Ie[e & 255] : 1792 <= e && e <= 2220 ? "AL" : "L";
}
function Te(e) {
  const t = e.length;
  if (t === 0)
    return null;
  const n = new Array(t);
  let u = 0;
  for (let s = 0; s < t; s++) {
    const o = Re(e.charCodeAt(s));
    (o === "R" || o === "AL" || o === "AN") && u++, n[s] = o;
  }
  if (u === 0)
    return null;
  const i = t / u < 0.3 ? 0 : 1, r = new Int8Array(t);
  for (let s = 0; s < t; s++)
    r[s] = i;
  const l = i & 1 ? "R" : "L", c = l;
  let a = c;
  for (let s = 0; s < t; s++)
    n[s] === "NSM" ? n[s] = a : a = n[s];
  a = c;
  for (let s = 0; s < t; s++) {
    const o = n[s];
    o === "EN" ? n[s] = a === "AL" ? "AN" : "EN" : (o === "R" || o === "L" || o === "AL") && (a = o);
  }
  for (let s = 0; s < t; s++)
    n[s] === "AL" && (n[s] = "R");
  for (let s = 1; s < t - 1; s++)
    n[s] === "ES" && n[s - 1] === "EN" && n[s + 1] === "EN" && (n[s] = "EN"), n[s] === "CS" && (n[s - 1] === "EN" || n[s - 1] === "AN") && n[s + 1] === n[s - 1] && (n[s] = n[s - 1]);
  for (let s = 0; s < t; s++) {
    if (n[s] !== "EN")
      continue;
    let o;
    for (o = s - 1; o >= 0 && n[o] === "ET"; o--)
      n[o] = "EN";
    for (o = s + 1; o < t && n[o] === "ET"; o++)
      n[o] = "EN";
  }
  for (let s = 0; s < t; s++) {
    const o = n[s];
    (o === "WS" || o === "ES" || o === "ET" || o === "CS") && (n[s] = "ON");
  }
  a = c;
  for (let s = 0; s < t; s++) {
    const o = n[s];
    o === "EN" ? n[s] = a === "L" ? "L" : "EN" : (o === "R" || o === "L") && (a = o);
  }
  for (let s = 0; s < t; s++) {
    if (n[s] !== "ON")
      continue;
    let o = s + 1;
    for (; o < t && n[o] === "ON"; )
      o++;
    const p = s > 0 ? n[s - 1] : c, f = o < t ? n[o] : c, h = p !== "L" ? "R" : "L";
    if (h === (f !== "L" ? "R" : "L"))
      for (let g = s; g < o; g++)
        n[g] = h;
    s = o - 1;
  }
  for (let s = 0; s < t; s++)
    n[s] === "ON" && (n[s] = l);
  for (let s = 0; s < t; s++) {
    const o = n[s];
    (r[s] & 1) === 0 ? o === "R" ? r[s]++ : (o === "AN" || o === "EN") && (r[s] += 2) : (o === "L" || o === "AN" || o === "EN") && r[s]++;
  }
  return r;
}
function Ge(e, t) {
  const n = Te(e);
  if (n === null)
    return null;
  const u = new Int8Array(t.length);
  for (let i = 0; i < t.length; i++)
    u[i] = n[t[i]];
  return u;
}
const je = /[ \t\n\r\f]+/g, ze = /[\t\n\r\f]| {2,}|^ | $/;
function He(e) {
  const t = e ?? "normal";
  return t === "pre-wrap" ? { mode: t, preserveOrdinarySpaces: !0, preserveHardBreaks: !0 } : { mode: t, preserveOrdinarySpaces: !1, preserveHardBreaks: !1 };
}
function De(e) {
  if (!ze.test(e))
    return e;
  let t = e.replace(je, " ");
  return t.charCodeAt(0) === 32 && (t = t.slice(1)), t.length > 0 && t.charCodeAt(t.length - 1) === 32 && (t = t.slice(0, -1)), t;
}
function Ke(e) {
  return /[\r\f]/.test(e) ? e.replace(/\r\n/g, `
`).replace(/[\r\f]/g, `
`) : e.replace(/\r\n/g, `
`);
}
let q = null, ie;
function Je() {
  return q === null && (q = new Intl.Segmenter(ie, { granularity: "word" })), q;
}
function Qe() {
  q = null;
}
function $e(e) {
  const t = e && e.length > 0 ? e : void 0;
  ie !== t && (ie = t, q = null);
}
const Ue = new RegExp("\\p{Script=Arabic}", "u"), te = new RegExp("\\p{M}", "u"), xe = new RegExp("\\p{Nd}", "u");
function se(e) {
  return Ue.test(e);
}
function Q(e) {
  for (const t of e) {
    const n = t.codePointAt(0);
    if (n >= 19968 && n <= 40959 || n >= 13312 && n <= 19903 || n >= 131072 && n <= 173791 || n >= 173824 && n <= 177983 || n >= 177984 && n <= 178207 || n >= 178208 && n <= 183983 || n >= 183984 && n <= 191471 || n >= 196608 && n <= 201551 || n >= 63744 && n <= 64255 || n >= 194560 && n <= 195103 || n >= 12288 && n <= 12351 || n >= 12352 && n <= 12447 || n >= 12448 && n <= 12543 || n >= 44032 && n <= 55215 || n >= 65280 && n <= 65519)
      return !0;
  }
  return !1;
}
const Ne = /* @__PURE__ */ new Set([
  "，",
  "．",
  "！",
  "：",
  "；",
  "？",
  "、",
  "。",
  "・",
  "）",
  "〕",
  "〉",
  "》",
  "」",
  "』",
  "】",
  "〗",
  "〙",
  "〛",
  "ー",
  "々",
  "〻",
  "ゝ",
  "ゞ",
  "ヽ",
  "ヾ"
]), ne = /* @__PURE__ */ new Set([
  '"',
  "(",
  "[",
  "{",
  "“",
  "‘",
  "«",
  "‹",
  "（",
  "〔",
  "〈",
  "《",
  "「",
  "『",
  "【",
  "〖",
  "〘",
  "〚"
]), ae = /* @__PURE__ */ new Set([
  "'",
  "’"
]), Z = /* @__PURE__ */ new Set([
  ".",
  ",",
  "!",
  "?",
  ":",
  ";",
  "،",
  "؛",
  "؟",
  "।",
  "॥",
  "၊",
  "။",
  "၌",
  "၍",
  "၏",
  ")",
  "]",
  "}",
  "%",
  '"',
  "”",
  "’",
  "»",
  "›",
  "…"
]), _e = /* @__PURE__ */ new Set([
  ":",
  ".",
  "،",
  "؛"
]), qe = /* @__PURE__ */ new Set([
  "၏"
]), Ze = /* @__PURE__ */ new Set([
  "”",
  "’",
  "»",
  "›",
  "」",
  "』",
  "】",
  "》",
  "〉",
  "〕",
  "）"
]);
function Ve(e) {
  if (he(e))
    return !0;
  let t = !1;
  for (const n of e) {
    if (Z.has(n)) {
      t = !0;
      continue;
    }
    if (!(t && te.test(n)))
      return !1;
  }
  return t;
}
function Xe(e) {
  for (const t of e)
    if (!Ne.has(t) && !Z.has(t))
      return !1;
  return e.length > 0;
}
function Ye(e) {
  if (he(e))
    return !0;
  for (const t of e)
    if (!ne.has(t) && !ae.has(t) && !te.test(t))
      return !1;
  return e.length > 0;
}
function he(e) {
  let t = !1;
  for (const n of e)
    if (!(n === "\\" || te.test(n))) {
      if (ne.has(n) || Z.has(n) || ae.has(n)) {
        t = !0;
        continue;
      }
      return !1;
    }
  return t;
}
function et(e) {
  const t = Array.from(e);
  let n = t.length;
  for (; n > 0; ) {
    const u = t[n - 1];
    if (te.test(u)) {
      n--;
      continue;
    }
    if (ne.has(u) || ae.has(u)) {
      n--;
      continue;
    }
    break;
  }
  return n <= 0 || n === t.length ? null : {
    head: t.slice(0, n).join(""),
    tail: t.slice(n).join("")
  };
}
function tt(e, t) {
  if (e.length === 0)
    return !1;
  for (const n of e)
    if (n !== t)
      return !1;
  return !0;
}
function nt(e) {
  return !se(e) || e.length === 0 ? !1 : _e.has(e[e.length - 1]);
}
function rt(e) {
  return e.length === 0 ? !1 : qe.has(e[e.length - 1]);
}
function it(e) {
  if (e.length < 2 || e[0] !== " ")
    return null;
  const t = e.slice(1);
  return new RegExp("^\\p{M}+$", "u").test(t) ? { space: " ", marks: t } : null;
}
function me(e) {
  for (let t = e.length - 1; t >= 0; t--) {
    const n = e[t];
    if (Ze.has(n))
      return !0;
    if (!Z.has(n))
      return !1;
  }
  return !1;
}
function st(e, t) {
  if (t.preserveOrdinarySpaces || t.preserveHardBreaks) {
    if (e === " ")
      return "preserved-space";
    if (e === "	")
      return "tab";
    if (t.preserveHardBreaks && e === `
`)
      return "hard-break";
  }
  return e === " " ? "space" : e === " " || e === " " || e === "⁠" || e === "\uFEFF" ? "glue" : e === "​" ? "zero-width-break" : e === "­" ? "soft-hyphen" : "text";
}
function H(e) {
  return e.length === 1 ? e[0] : e.join("");
}
function ut(e, t, n, u) {
  const i = [];
  let r = null, l = [], c = n, a = !1, s = 0;
  for (const o of e) {
    const p = st(o, u), f = p === "text" && t;
    if (r !== null && p === r && f === a) {
      l.push(o), s += o.length;
      continue;
    }
    r !== null && i.push({
      text: H(l),
      isWordLike: a,
      kind: r,
      start: c
    }), r = p, l = [o], c = n + s, a = f, s += o.length;
  }
  return r !== null && i.push({
    text: H(l),
    isWordLike: a,
    kind: r,
    start: c
  }), i;
}
function ue(e) {
  return e === "space" || e === "preserved-space" || e === "zero-width-break" || e === "hard-break";
}
const lt = /^[A-Za-z][A-Za-z0-9+.-]*:$/;
function ot(e, t) {
  const n = e.texts[t];
  return n.startsWith("www.") ? !0 : lt.test(n) && t + 1 < e.len && e.kinds[t + 1] === "text" && e.texts[t + 1] === "//";
}
function ct(e) {
  return e.includes("?") && (e.includes("://") || e.startsWith("www."));
}
function ft(e) {
  const t = e.texts.slice(), n = e.isWordLike.slice(), u = e.kinds.slice(), i = e.starts.slice();
  for (let l = 0; l < e.len; l++) {
    if (u[l] !== "text" || !ot(e, l))
      continue;
    const c = [t[l]];
    let a = l + 1;
    for (; a < e.len && !ue(u[a]); ) {
      c.push(t[a]), n[l] = !0;
      const s = t[a].includes("?");
      if (u[a] = "text", t[a] = "", a++, s)
        break;
    }
    t[l] = H(c);
  }
  let r = 0;
  for (let l = 0; l < t.length; l++) {
    const c = t[l];
    c.length !== 0 && (r !== l && (t[r] = c, n[r] = n[l], u[r] = u[l], i[r] = i[l]), r++);
  }
  return t.length = r, n.length = r, u.length = r, i.length = r, {
    len: r,
    texts: t,
    isWordLike: n,
    kinds: u,
    starts: i
  };
}
function at(e) {
  const t = [], n = [], u = [], i = [];
  for (let r = 0; r < e.len; r++) {
    const l = e.texts[r];
    if (t.push(l), n.push(e.isWordLike[r]), u.push(e.kinds[r]), i.push(e.starts[r]), !ct(l))
      continue;
    const c = r + 1;
    if (c >= e.len || ue(e.kinds[c]))
      continue;
    const a = [], s = e.starts[c];
    let o = c;
    for (; o < e.len && !ue(e.kinds[o]); )
      a.push(e.texts[o]), o++;
    a.length > 0 && (t.push(H(a)), n.push(!0), u.push("text"), i.push(s), r = o - 1);
  }
  return {
    len: t.length,
    texts: t,
    isWordLike: n,
    kinds: u,
    starts: i
  };
}
const ht = /* @__PURE__ */ new Set([
  ":",
  "-",
  "/",
  "×",
  ",",
  ".",
  "+",
  "–",
  "—"
]), ge = /^[A-Za-z0-9_]+[,:;]*$/, Se = /[,:;]+$/;
function Be(e) {
  for (const t of e)
    if (xe.test(t))
      return !0;
  return !1;
}
function le(e) {
  if (e.length === 0)
    return !1;
  for (const t of e)
    if (!(xe.test(t) || ht.has(t)))
      return !1;
  return !0;
}
function dt(e) {
  const t = [], n = [], u = [], i = [];
  for (let r = 0; r < e.len; r++) {
    const l = e.texts[r], c = e.kinds[r];
    if (c === "text" && le(l) && Be(l)) {
      const a = [l];
      let s = r + 1;
      for (; s < e.len && e.kinds[s] === "text" && le(e.texts[s]); )
        a.push(e.texts[s]), s++;
      t.push(H(a)), n.push(!0), u.push("text"), i.push(e.starts[r]), r = s - 1;
      continue;
    }
    t.push(l), n.push(e.isWordLike[r]), u.push(c), i.push(e.starts[r]);
  }
  return {
    len: t.length,
    texts: t,
    isWordLike: n,
    kinds: u,
    starts: i
  };
}
function Lt(e) {
  const t = [], n = [], u = [], i = [];
  for (let r = 0; r < e.len; r++) {
    const l = e.texts[r], c = e.kinds[r], a = e.isWordLike[r];
    if (c === "text" && a && ge.test(l)) {
      const s = [l];
      let o = Se.test(l), p = r + 1;
      for (; o && p < e.len && e.kinds[p] === "text" && e.isWordLike[p] && ge.test(e.texts[p]); ) {
        const f = e.texts[p];
        s.push(f), o = Se.test(f), p++;
      }
      t.push(H(s)), n.push(!0), u.push("text"), i.push(e.starts[r]), r = p - 1;
      continue;
    }
    t.push(l), n.push(a), u.push(c), i.push(e.starts[r]);
  }
  return {
    len: t.length,
    texts: t,
    isWordLike: n,
    kinds: u,
    starts: i
  };
}
function pt(e) {
  const t = [], n = [], u = [], i = [];
  for (let r = 0; r < e.len; r++) {
    const l = e.texts[r];
    if (e.kinds[r] === "text" && l.includes("-")) {
      const c = l.split("-");
      let a = c.length > 1;
      for (let s = 0; s < c.length; s++) {
        const o = c[s];
        if (!a)
          break;
        (o.length === 0 || !Be(o) || !le(o)) && (a = !1);
      }
      if (a) {
        let s = 0;
        for (let o = 0; o < c.length; o++) {
          const p = c[o], f = o < c.length - 1 ? `${p}-` : p;
          t.push(f), n.push(!0), u.push("text"), i.push(e.starts[r] + s), s += f.length;
        }
        continue;
      }
    }
    t.push(l), n.push(e.isWordLike[r]), u.push(e.kinds[r]), i.push(e.starts[r]);
  }
  return {
    len: t.length,
    texts: t,
    isWordLike: n,
    kinds: u,
    starts: i
  };
}
function At(e) {
  const t = [], n = [], u = [], i = [];
  let r = 0;
  for (; r < e.len; ) {
    const l = [e.texts[r]];
    let c = e.isWordLike[r], a = e.kinds[r], s = e.starts[r];
    if (a === "glue") {
      const o = [l[0]], p = s;
      for (r++; r < e.len && e.kinds[r] === "glue"; )
        o.push(e.texts[r]), r++;
      const f = H(o);
      if (r < e.len && e.kinds[r] === "text")
        l[0] = f, l.push(e.texts[r]), c = e.isWordLike[r], a = "text", s = p, r++;
      else {
        t.push(f), n.push(!1), u.push("glue"), i.push(p);
        continue;
      }
    } else
      r++;
    if (a === "text")
      for (; r < e.len && e.kinds[r] === "glue"; ) {
        const o = [];
        for (; r < e.len && e.kinds[r] === "glue"; )
          o.push(e.texts[r]), r++;
        const p = H(o);
        if (r < e.len && e.kinds[r] === "text") {
          l.push(p, e.texts[r]), c = c || e.isWordLike[r], r++;
          continue;
        }
        l.push(p);
      }
    t.push(H(l)), n.push(c), u.push(a), i.push(s);
  }
  return {
    len: t.length,
    texts: t,
    isWordLike: n,
    kinds: u,
    starts: i
  };
}
function gt(e) {
  const t = e.texts.slice(), n = e.isWordLike.slice(), u = e.kinds.slice(), i = e.starts.slice();
  for (let r = 0; r < t.length - 1; r++) {
    if (u[r] !== "text" || u[r + 1] !== "text" || !Q(t[r]) || !Q(t[r + 1]))
      continue;
    const l = et(t[r]);
    l !== null && (t[r] = l.head, t[r + 1] = l.tail + t[r + 1], i[r + 1] = i[r] + l.head.length);
  }
  return {
    len: t.length,
    texts: t,
    isWordLike: n,
    kinds: u,
    starts: i
  };
}
function St(e, t, n) {
  const u = Je();
  let i = 0;
  const r = [], l = [], c = [], a = [];
  for (const f of u.segment(e))
    for (const h of ut(f.segment, f.isWordLike ?? !1, f.index, n)) {
      const W = h.kind === "text";
      t.carryCJKAfterClosingQuote && W && i > 0 && c[i - 1] === "text" && Q(h.text) && Q(r[i - 1]) && me(r[i - 1]) || W && i > 0 && c[i - 1] === "text" && Xe(h.text) && Q(r[i - 1]) || W && i > 0 && c[i - 1] === "text" && rt(r[i - 1]) ? (r[i - 1] += h.text, l[i - 1] = l[i - 1] || h.isWordLike) : W && i > 0 && c[i - 1] === "text" && h.isWordLike && se(h.text) && nt(r[i - 1]) ? (r[i - 1] += h.text, l[i - 1] = !0) : W && !h.isWordLike && i > 0 && c[i - 1] === "text" && h.text.length === 1 && h.text !== "-" && h.text !== "—" && tt(r[i - 1], h.text) || W && !h.isWordLike && i > 0 && c[i - 1] === "text" && (Ve(h.text) || h.text === "-" && l[i - 1]) ? r[i - 1] += h.text : (r[i] = h.text, l[i] = h.isWordLike, c[i] = h.kind, a[i] = h.start, i++);
    }
  for (let f = 1; f < i; f++)
    c[f] === "text" && !l[f] && he(r[f]) && c[f - 1] === "text" && (r[f - 1] += r[f], l[f - 1] = l[f - 1] || l[f], r[f] = "");
  for (let f = i - 2; f >= 0; f--)
    if (c[f] === "text" && !l[f] && Ye(r[f])) {
      let h = f + 1;
      for (; h < i && r[h] === ""; )
        h++;
      h < i && c[h] === "text" && (r[h] = r[f] + r[h], a[h] = a[f], r[f] = "");
    }
  let s = 0;
  for (let f = 0; f < i; f++) {
    const h = r[f];
    h.length !== 0 && (s !== f && (r[s] = h, l[s] = l[f], c[s] = c[f], a[s] = a[f]), s++);
  }
  r.length = s, l.length = s, c.length = s, a.length = s;
  const o = At({
    len: s,
    texts: r,
    isWordLike: l,
    kinds: c,
    starts: a
  }), p = gt(Lt(pt(dt(at(ft(o))))));
  for (let f = 0; f < p.len - 1; f++) {
    const h = it(p.texts[f]);
    h !== null && (p.kinds[f] !== "space" && p.kinds[f] !== "preserved-space" || p.kinds[f + 1] !== "text" || !se(p.texts[f + 1]) || (p.texts[f] = h.space, p.isWordLike[f] = !1, p.kinds[f] = p.kinds[f] === "preserved-space" ? "preserved-space" : "space", p.texts[f + 1] = h.marks + p.texts[f + 1], p.starts[f + 1] = p.starts[f] + h.space.length));
  }
  return p;
}
function kt(e, t) {
  if (e.len === 0)
    return [];
  if (!t.preserveHardBreaks)
    return [{
      startSegmentIndex: 0,
      endSegmentIndex: e.len,
      consumedEndSegmentIndex: e.len
    }];
  const n = [];
  let u = 0;
  for (let i = 0; i < e.len; i++)
    e.kinds[i] === "hard-break" && (n.push({
      startSegmentIndex: u,
      endSegmentIndex: i,
      consumedEndSegmentIndex: i + 1
    }), u = i + 1);
  return u < e.len && n.push({
    startSegmentIndex: u,
    endSegmentIndex: e.len,
    consumedEndSegmentIndex: e.len
  }), n;
}
function We(e, t, n = "normal") {
  const u = He(n), i = u.mode === "pre-wrap" ? Ke(e) : De(e);
  if (i.length === 0)
    return {
      normalized: i,
      chunks: [],
      len: 0,
      texts: [],
      isWordLike: [],
      kinds: [],
      starts: []
    };
  const r = St(i, t, u);
  return {
    normalized: i,
    chunks: kt(r, u),
    ...r
  };
}
let U = null;
const oe = /* @__PURE__ */ new Map();
let _ = null;
const xt = new RegExp("\\p{Emoji_Presentation}", "u"), Nt = /[\p{Emoji_Presentation}\p{Extended_Pictographic}\p{Regional_Indicator}\uFE0F\u20E3]/u;
let Y = null;
const ce = /* @__PURE__ */ new Map();
function de() {
  if (U !== null)
    return U;
  if (typeof OffscreenCanvas < "u")
    return U = new OffscreenCanvas(1, 1).getContext("2d"), U;
  if (typeof document < "u")
    return U = document.createElement("canvas").getContext("2d"), U;
  throw new Error("Text measurement requires OffscreenCanvas or a DOM canvas context.");
}
function mt(e) {
  let t = oe.get(e);
  return t || (t = /* @__PURE__ */ new Map(), oe.set(e, t)), t;
}
function K(e, t) {
  let n = t.get(e);
  return n === void 0 && (n = {
    width: de().measureText(e).width,
    containsCJK: Q(e)
  }, t.set(e, n)), n;
}
function $() {
  if (_ !== null)
    return _;
  if (typeof navigator > "u")
    return _ = {
      lineFitEpsilon: 5e-3,
      carryCJKAfterClosingQuote: !1,
      preferPrefixWidthsForBreakableRuns: !1,
      preferEarlySoftHyphenBreak: !1
    }, _;
  const e = navigator.userAgent, n = navigator.vendor === "Apple Computer, Inc." && e.includes("Safari/") && !e.includes("Chrome/") && !e.includes("Chromium/") && !e.includes("CriOS/") && !e.includes("FxiOS/") && !e.includes("EdgiOS/"), u = e.includes("Chrome/") || e.includes("Chromium/") || e.includes("CriOS/") || e.includes("Edg/");
  return _ = {
    lineFitEpsilon: n ? 1 / 64 : 5e-3,
    carryCJKAfterClosingQuote: u,
    preferPrefixWidthsForBreakableRuns: n,
    preferEarlySoftHyphenBreak: n
  }, _;
}
function Bt(e) {
  const t = e.match(/(\d+(?:\.\d+)?)\s*px/);
  return t ? parseFloat(t[1]) : 16;
}
function Le() {
  return Y === null && (Y = new Intl.Segmenter(void 0, { granularity: "grapheme" })), Y;
}
function Wt(e) {
  return xt.test(e) || e.includes("️");
}
function Et(e) {
  return Nt.test(e);
}
function wt(e, t) {
  let n = ce.get(e);
  if (n !== void 0)
    return n;
  const u = de();
  u.font = e;
  const i = u.measureText("😀").width;
  if (n = 0, i > t + 0.5 && typeof document < "u" && document.body !== null) {
    const r = document.createElement("span");
    r.style.font = e, r.style.display = "inline-block", r.style.visibility = "hidden", r.style.position = "absolute", r.textContent = "😀", document.body.appendChild(r);
    const l = r.getBoundingClientRect().width;
    document.body.removeChild(r), i - l > 0.5 && (n = i - l);
  }
  return ce.set(e, n), n;
}
function Ft(e) {
  let t = 0;
  const n = Le();
  for (const u of n.segment(e))
    Wt(u.segment) && t++;
  return t;
}
function bt(e, t) {
  return t.emojiCount === void 0 && (t.emojiCount = Ft(e)), t.emojiCount;
}
function J(e, t, n) {
  return n === 0 ? t.width : t.width - bt(e, t) * n;
}
function Ct(e, t, n, u) {
  if (t.graphemeWidths !== void 0)
    return t.graphemeWidths;
  const i = [], r = Le();
  for (const l of r.segment(e)) {
    const c = K(l.segment, n);
    i.push(J(l.segment, c, u));
  }
  return t.graphemeWidths = i.length > 1 ? i : null, t.graphemeWidths;
}
function Pt(e, t, n, u) {
  if (t.graphemePrefixWidths !== void 0)
    return t.graphemePrefixWidths;
  const i = [], r = Le();
  let l = "";
  for (const c of r.segment(e)) {
    l += c.segment;
    const a = K(l, n);
    i.push(J(l, a, u));
  }
  return t.graphemePrefixWidths = i.length > 1 ? i : null, t.graphemePrefixWidths;
}
function yt(e, t) {
  const n = de();
  n.font = e;
  const u = mt(e), i = Bt(e), r = t ? wt(e, i) : 0;
  return { cache: u, fontSize: i, emojiCorrection: r };
}
function Mt() {
  oe.clear(), ce.clear(), Y = null;
}
function D(e) {
  return e === "space" || e === "preserved-space" || e === "tab" || e === "zero-width-break" || e === "soft-hyphen";
}
function Ot(e, t) {
  for (; t < e.widths.length; ) {
    const n = e.kinds[t];
    if (n !== "space" && n !== "zero-width-break" && n !== "soft-hyphen")
      break;
    t++;
  }
  return t;
}
function Ee(e, t) {
  if (t <= 0)
    return 0;
  const n = e % t;
  return Math.abs(n) <= 1e-6 ? t : t - n;
}
function re(e, t, n, u) {
  return !u || t === null ? e[n] : t[n] - (n > 0 ? t[n - 1] : 0);
}
function we(e, t, n, u, i, r) {
  let l = 0, c = t;
  for (; l < e.length; ) {
    const a = r ? t + e[l] : c + e[l];
    if ((l + 1 < e.length ? a + i : a) > n + u)
      break;
    c = a, l++;
  }
  return { fitCount: l, fittedWidth: c };
}
function vt(e, t) {
  let n = 0, u = e.chunks.length;
  for (; n < u; ) {
    const i = Math.floor((n + u) / 2);
    t < e.chunks[i].consumedEndSegmentIndex ? u = i : n = i + 1;
  }
  return n < e.chunks.length ? n : -1;
}
function It(e, t) {
  let n = t.segmentIndex;
  const u = t.graphemeIndex;
  if (n >= e.widths.length)
    return null;
  const i = vt(e, n);
  if (i < 0)
    return null;
  if (u > 0)
    return { cursor: t, chunkIndex: i };
  const r = e.chunks[i];
  if (r.startSegmentIndex === r.endSegmentIndex && n === r.startSegmentIndex)
    return { cursor: { segmentIndex: n, graphemeIndex: 0 }, chunkIndex: i };
  for (n < r.startSegmentIndex && (n = r.startSegmentIndex); n < r.endSegmentIndex; ) {
    const l = e.kinds[n];
    if (l !== "space" && l !== "zero-width-break" && l !== "soft-hyphen")
      return { cursor: { segmentIndex: n, graphemeIndex: 0 }, chunkIndex: i };
    n++;
  }
  return r.consumedEndSegmentIndex >= e.widths.length ? null : {
    cursor: { segmentIndex: r.consumedEndSegmentIndex, graphemeIndex: 0 },
    chunkIndex: i + 1
  };
}
function Rt(e, t) {
  return e.simpleLineWalkFastPath ? Tt(e, t) : pe(e, t);
}
function Tt(e, t) {
  return Fe(e, t);
}
function Fe(e, t, n) {
  const { widths: u, kinds: i, breakableWidths: r, breakablePrefixWidths: l } = e;
  if (u.length === 0)
    return 0;
  const c = $(), a = c.lineFitEpsilon;
  let s = 0, o = 0, p = !1, f = 0, h = 0, W = 0, g = 0, C = -1, v = 0;
  function T() {
    C = -1, v = 0;
  }
  function E(A = W, B = g, M = o) {
    s++, n == null || n({
      startSegmentIndex: f,
      startGraphemeIndex: h,
      endSegmentIndex: A,
      endGraphemeIndex: B,
      width: M
    }), o = 0, p = !1, T();
  }
  function w(A, B) {
    p = !0, f = A, h = 0, W = A + 1, g = 0, o = B;
  }
  function b(A, B, M) {
    p = !0, f = A, h = B, W = A, g = B + 1, o = M;
  }
  function I(A, B) {
    if (!p) {
      w(A, B);
      return;
    }
    o += B, W = A + 1, g = 0;
  }
  function k(A, B) {
    D(i[A]) && (C = A + 1, v = o - B);
  }
  function x(A) {
    y(A, 0);
  }
  function y(A, B) {
    const M = r[A], G = l[A] ?? null;
    for (let R = B; R < M.length; R++) {
      const j = re(M, G, R, c.preferPrefixWidthsForBreakableRuns);
      if (!p) {
        b(A, R, j);
        continue;
      }
      o + j > t + a ? (E(), b(A, R, j)) : (o += j, W = A, g = R + 1);
    }
    p && W === A && g === M.length && (W = A + 1, g = 0);
  }
  let N = 0;
  for (; N < u.length && !(!p && (N = Ot(e, N), N >= u.length)); ) {
    const A = u[N], B = i[N];
    if (!p) {
      A > t && r[N] !== null ? x(N) : w(N, A), k(N, A), N++;
      continue;
    }
    if (o + A > t + a) {
      if (D(B)) {
        I(N, A), E(N + 1, 0, o - A), N++;
        continue;
      }
      if (C >= 0) {
        if (W > C || W === C && g > 0) {
          E();
          continue;
        }
        E(C, 0, v);
        continue;
      }
      if (A > t && r[N] !== null) {
        E(), x(N), N++;
        continue;
      }
      E();
      continue;
    }
    I(N, A), k(N, A), N++;
  }
  return p && E(), s;
}
function pe(e, t, n) {
  if (e.simpleLineWalkFastPath)
    return Fe(e, t, n);
  const { widths: u, lineEndFitAdvances: i, lineEndPaintAdvances: r, kinds: l, breakableWidths: c, breakablePrefixWidths: a, discretionaryHyphenWidth: s, tabStopAdvance: o, chunks: p } = e;
  if (u.length === 0 || p.length === 0)
    return 0;
  const f = $(), h = f.lineFitEpsilon;
  let W = 0, g = 0, C = !1, v = 0, T = 0, E = 0, w = 0, b = -1, I = 0, k = 0, x = null;
  function y() {
    b = -1, I = 0, k = 0, x = null;
  }
  function N(L = E, S = w, m = g) {
    W++, n == null || n({
      startSegmentIndex: v,
      startGraphemeIndex: T,
      endSegmentIndex: L,
      endGraphemeIndex: S,
      width: m
    }), g = 0, C = !1, y();
  }
  function A(L, S) {
    C = !0, v = L, T = 0, E = L + 1, w = 0, g = S;
  }
  function B(L, S, m) {
    C = !0, v = L, T = S, E = L, w = S + 1, g = m;
  }
  function M(L, S) {
    if (!C) {
      A(L, S);
      return;
    }
    g += S, E = L + 1, w = 0;
  }
  function G(L, S) {
    if (!D(l[L]))
      return;
    const m = l[L] === "tab" ? 0 : i[L], P = l[L] === "tab" ? S : r[L];
    b = L + 1, I = g - S + m, k = g - S + P, x = l[L];
  }
  function R(L) {
    j(L, 0);
  }
  function j(L, S) {
    const m = c[L], P = a[L] ?? null;
    for (let O = S; O < m.length; O++) {
      const z = re(m, P, O, f.preferPrefixWidthsForBreakableRuns);
      if (!C) {
        B(L, O, z);
        continue;
      }
      g + z > t + h ? (N(), B(L, O, z)) : (g += z, E = L, w = O + 1);
    }
    C && E === L && w === m.length && (E = L + 1, w = 0);
  }
  function d(L) {
    if (x !== "soft-hyphen")
      return !1;
    const S = c[L];
    if (S === null)
      return !1;
    const m = f.preferPrefixWidthsForBreakableRuns ? a[L] ?? S : S, P = m !== S, { fitCount: O, fittedWidth: z } = we(m, g, t, h, s, P);
    return O === 0 ? !1 : (g = z, E = L, w = O, y(), O === S.length ? (E = L + 1, w = 0, !0) : (N(L, O, z + s), j(L, O), !0));
  }
  function F(L) {
    W++, n == null || n({
      startSegmentIndex: L.startSegmentIndex,
      startGraphemeIndex: 0,
      endSegmentIndex: L.consumedEndSegmentIndex,
      endGraphemeIndex: 0,
      width: 0
    }), y();
  }
  for (let L = 0; L < p.length; L++) {
    const S = p[L];
    if (S.startSegmentIndex === S.endSegmentIndex) {
      F(S);
      continue;
    }
    C = !1, g = 0, v = S.startSegmentIndex, T = 0, E = S.startSegmentIndex, w = 0, y();
    let m = S.startSegmentIndex;
    for (; m < S.endSegmentIndex; ) {
      const P = l[m], O = P === "tab" ? Ee(g, o) : u[m];
      if (P === "soft-hyphen") {
        C && (E = m + 1, w = 0, b = m + 1, I = g + s, k = g + s, x = P), m++;
        continue;
      }
      if (!C) {
        O > t && c[m] !== null ? R(m) : A(m, O), G(m, O), m++;
        continue;
      }
      if (g + O > t + h) {
        const V = g + (P === "tab" ? 0 : i[m]), X = g + (P === "tab" ? O : r[m]);
        if (x === "soft-hyphen" && f.preferEarlySoftHyphenBreak && I <= t + h) {
          N(b, 0, k);
          continue;
        }
        if (x === "soft-hyphen" && d(m)) {
          m++;
          continue;
        }
        if (D(P) && V <= t + h) {
          M(m, O), N(m + 1, 0, X), m++;
          continue;
        }
        if (b >= 0 && I <= t + h) {
          if (E > b || E === b && w > 0) {
            N();
            continue;
          }
          const Ae = b;
          N(Ae, 0, k), m = Ae;
          continue;
        }
        if (O > t && c[m] !== null) {
          N(), R(m), m++;
          continue;
        }
        N();
        continue;
      }
      M(m, O), G(m, O), m++;
    }
    if (C) {
      const P = b === S.consumedEndSegmentIndex ? k : g;
      N(S.consumedEndSegmentIndex, 0, P);
    }
  }
  return W;
}
function Gt(e, t, n) {
  const u = It(e, t);
  if (u === null)
    return null;
  if (e.simpleLineWalkFastPath)
    return jt(e, u.cursor, n);
  const i = e.chunks[u.chunkIndex];
  if (i.startSegmentIndex === i.endSegmentIndex)
    return {
      startSegmentIndex: i.startSegmentIndex,
      startGraphemeIndex: 0,
      endSegmentIndex: i.consumedEndSegmentIndex,
      endGraphemeIndex: 0,
      width: 0
    };
  const { widths: r, lineEndFitAdvances: l, lineEndPaintAdvances: c, kinds: a, breakableWidths: s, breakablePrefixWidths: o, discretionaryHyphenWidth: p, tabStopAdvance: f } = e, h = $(), W = h.lineFitEpsilon;
  let g = 0, C = !1;
  const v = u.cursor.segmentIndex, T = u.cursor.graphemeIndex;
  let E = v, w = T, b = -1, I = 0, k = 0, x = null;
  function y() {
    b = -1, I = 0, k = 0, x = null;
  }
  function N(d = E, F = w, L = g) {
    return C ? {
      startSegmentIndex: v,
      startGraphemeIndex: T,
      endSegmentIndex: d,
      endGraphemeIndex: F,
      width: L
    } : null;
  }
  function A(d, F) {
    C = !0, E = d + 1, w = 0, g = F;
  }
  function B(d, F, L) {
    C = !0, E = d, w = F + 1, g = L;
  }
  function M(d, F) {
    if (!C) {
      A(d, F);
      return;
    }
    g += F, E = d + 1, w = 0;
  }
  function G(d, F) {
    if (!D(a[d]))
      return;
    const L = a[d] === "tab" ? 0 : l[d], S = a[d] === "tab" ? F : c[d];
    b = d + 1, I = g - F + L, k = g - F + S, x = a[d];
  }
  function R(d, F) {
    const L = s[d], S = o[d] ?? null;
    for (let m = F; m < L.length; m++) {
      const P = re(L, S, m, h.preferPrefixWidthsForBreakableRuns);
      if (!C) {
        B(d, m, P);
        continue;
      }
      if (g + P > n + W)
        return N();
      g += P, E = d, w = m + 1;
    }
    return C && E === d && w === L.length && (E = d + 1, w = 0), null;
  }
  function j(d) {
    if (x !== "soft-hyphen" || b < 0)
      return null;
    const F = s[d] ?? null;
    if (F !== null) {
      const L = h.preferPrefixWidthsForBreakableRuns ? o[d] ?? F : F, S = L !== F, { fitCount: m, fittedWidth: P } = we(L, g, n, W, p, S);
      if (m === F.length)
        return g = P, E = d + 1, w = 0, y(), null;
      if (m > 0)
        return N(d, m, P + p);
    }
    return I <= n + W ? N(b, 0, k) : null;
  }
  for (let d = u.cursor.segmentIndex; d < i.endSegmentIndex; d++) {
    const F = a[d], L = d === u.cursor.segmentIndex ? u.cursor.graphemeIndex : 0, S = F === "tab" ? Ee(g, f) : r[d];
    if (F === "soft-hyphen" && L === 0) {
      C && (E = d + 1, w = 0, b = d + 1, I = g + p, k = g + p, x = F);
      continue;
    }
    if (!C) {
      if (L > 0) {
        const P = R(d, L);
        if (P !== null)
          return P;
      } else if (S > n && s[d] !== null) {
        const P = R(d, 0);
        if (P !== null)
          return P;
      } else
        A(d, S);
      G(d, S);
      continue;
    }
    if (g + S > n + W) {
      const P = g + (F === "tab" ? 0 : l[d]), O = g + (F === "tab" ? S : c[d]);
      if (x === "soft-hyphen" && h.preferEarlySoftHyphenBreak && I <= n + W)
        return N(b, 0, k);
      const z = j(d);
      if (z !== null)
        return z;
      if (D(F) && P <= n + W)
        return M(d, S), N(d + 1, 0, O);
      if (b >= 0 && I <= n + W)
        return E > b || E === b && w > 0 ? N() : N(b, 0, k);
      if (S > n && s[d] !== null) {
        const V = N();
        if (V !== null)
          return V;
        const X = R(d, 0);
        if (X !== null)
          return X;
      }
      return N();
    }
    M(d, S), G(d, S);
  }
  return b === i.consumedEndSegmentIndex && w === 0 ? N(i.consumedEndSegmentIndex, 0, k) : N(i.consumedEndSegmentIndex, 0, g);
}
function jt(e, t, n) {
  const { widths: u, kinds: i, breakableWidths: r, breakablePrefixWidths: l } = e, c = $(), a = c.lineFitEpsilon;
  let s = 0, o = !1;
  const p = t.segmentIndex, f = t.graphemeIndex;
  let h = p, W = f, g = -1, C = 0;
  function v(k = h, x = W, y = s) {
    return o ? {
      startSegmentIndex: p,
      startGraphemeIndex: f,
      endSegmentIndex: k,
      endGraphemeIndex: x,
      width: y
    } : null;
  }
  function T(k, x) {
    o = !0, h = k + 1, W = 0, s = x;
  }
  function E(k, x, y) {
    o = !0, h = k, W = x + 1, s = y;
  }
  function w(k, x) {
    if (!o) {
      T(k, x);
      return;
    }
    s += x, h = k + 1, W = 0;
  }
  function b(k, x) {
    D(i[k]) && (g = k + 1, C = s - x);
  }
  function I(k, x) {
    const y = r[k], N = l[k] ?? null;
    for (let A = x; A < y.length; A++) {
      const B = re(y, N, A, c.preferPrefixWidthsForBreakableRuns);
      if (!o) {
        E(k, A, B);
        continue;
      }
      if (s + B > n + a)
        return v();
      s += B, h = k, W = A + 1;
    }
    return o && h === k && W === y.length && (h = k + 1, W = 0), null;
  }
  for (let k = t.segmentIndex; k < u.length; k++) {
    const x = u[k], y = i[k], N = k === t.segmentIndex ? t.graphemeIndex : 0;
    if (!o) {
      if (N > 0) {
        const B = I(k, N);
        if (B !== null)
          return B;
      } else if (x > n && r[k] !== null) {
        const B = I(k, 0);
        if (B !== null)
          return B;
      } else
        T(k, x);
      b(k, x);
      continue;
    }
    if (s + x > n + a) {
      if (D(y))
        return w(k, x), v(k + 1, 0, s - x);
      if (g >= 0)
        return h > g || h === g && W > 0 ? v() : v(g, 0, C);
      if (x > n && r[k] !== null) {
        const B = v();
        if (B !== null)
          return B;
        const M = I(k, 0);
        if (M !== null)
          return M;
      }
      return v();
    }
    w(k, x), b(k, x);
  }
  return v();
}
let ee = null, fe = /* @__PURE__ */ new WeakMap();
function be() {
  return ee === null && (ee = new Intl.Segmenter(void 0, { granularity: "grapheme" })), ee;
}
function zt(e) {
  return e ? {
    widths: [],
    lineEndFitAdvances: [],
    lineEndPaintAdvances: [],
    kinds: [],
    simpleLineWalkFastPath: !0,
    segLevels: null,
    breakableWidths: [],
    breakablePrefixWidths: [],
    discretionaryHyphenWidth: 0,
    tabStopAdvance: 0,
    chunks: [],
    segments: []
  } : {
    widths: [],
    lineEndFitAdvances: [],
    lineEndPaintAdvances: [],
    kinds: [],
    simpleLineWalkFastPath: !0,
    segLevels: null,
    breakableWidths: [],
    breakablePrefixWidths: [],
    discretionaryHyphenWidth: 0,
    tabStopAdvance: 0,
    chunks: []
  };
}
function Ce(e, t, n) {
  const u = be(), i = $(), { cache: r, emojiCorrection: l } = yt(t, Et(e.normalized)), c = J("-", K("-", r), l), s = J(" ", K(" ", r), l) * 8;
  if (e.len === 0)
    return zt(n);
  const o = [], p = [], f = [], h = [];
  let W = e.chunks.length <= 1;
  const g = n ? [] : null, C = [], v = [], T = n ? [] : null, E = Array.from({ length: e.len }), w = Array.from({ length: e.len });
  function b(x, y, N, A, B, M, G, R) {
    B !== "text" && B !== "space" && B !== "zero-width-break" && (W = !1), o.push(y), p.push(N), f.push(A), h.push(B), g == null || g.push(M), C.push(G), v.push(R), T !== null && T.push(x);
  }
  for (let x = 0; x < e.len; x++) {
    E[x] = o.length;
    const y = e.texts[x], N = e.isWordLike[x], A = e.kinds[x], B = e.starts[x];
    if (A === "soft-hyphen") {
      b(y, 0, c, c, A, B, null, null), w[x] = o.length;
      continue;
    }
    if (A === "hard-break") {
      b(y, 0, 0, 0, A, B, null, null), w[x] = o.length;
      continue;
    }
    if (A === "tab") {
      b(y, 0, 0, 0, A, B, null, null), w[x] = o.length;
      continue;
    }
    const M = K(y, r);
    if (A === "text" && M.containsCJK) {
      let d = "", F = 0;
      for (const L of u.segment(y)) {
        const S = L.segment;
        if (d.length === 0) {
          d = S, F = L.index;
          continue;
        }
        if (ne.has(d) || Ne.has(S) || Z.has(S) || i.carryCJKAfterClosingQuote && Q(S) && me(d)) {
          d += S;
          continue;
        }
        const m = K(d, r), P = J(d, m, l);
        b(d, P, P, P, "text", B + F, null, null), d = S, F = L.index;
      }
      if (d.length > 0) {
        const L = K(d, r), S = J(d, L, l);
        b(d, S, S, S, "text", B + F, null, null);
      }
      w[x] = o.length;
      continue;
    }
    const G = J(y, M, l), R = A === "space" || A === "preserved-space" || A === "zero-width-break" ? 0 : G, j = A === "space" || A === "zero-width-break" ? 0 : G;
    if (N && y.length > 1) {
      const d = Ct(y, M, r, l), F = i.preferPrefixWidthsForBreakableRuns ? Pt(y, M, r, l) : null;
      b(y, G, R, j, A, B, d, F);
    } else
      b(y, G, R, j, A, B, null, null);
    w[x] = o.length;
  }
  const I = Ht(e.chunks, E, w), k = g === null ? null : Ge(e.normalized, g);
  return T !== null ? {
    widths: o,
    lineEndFitAdvances: p,
    lineEndPaintAdvances: f,
    kinds: h,
    simpleLineWalkFastPath: W,
    segLevels: k,
    breakableWidths: C,
    breakablePrefixWidths: v,
    discretionaryHyphenWidth: c,
    tabStopAdvance: s,
    chunks: I,
    segments: T
  } : {
    widths: o,
    lineEndFitAdvances: p,
    lineEndPaintAdvances: f,
    kinds: h,
    simpleLineWalkFastPath: W,
    segLevels: k,
    breakableWidths: C,
    breakablePrefixWidths: v,
    discretionaryHyphenWidth: c,
    tabStopAdvance: s,
    chunks: I
  };
}
function Ht(e, t, n) {
  const u = [];
  for (let i = 0; i < e.length; i++) {
    const r = e[i], l = r.startSegmentIndex < t.length ? t[r.startSegmentIndex] : n[n.length - 1] ?? 0, c = r.endSegmentIndex < t.length ? t[r.endSegmentIndex] : n[n.length - 1] ?? 0, a = r.consumedEndSegmentIndex < t.length ? t[r.consumedEndSegmentIndex] : n[n.length - 1] ?? 0;
    u.push({
      startSegmentIndex: l,
      endSegmentIndex: c,
      consumedEndSegmentIndex: a
    });
  }
  return u;
}
function Pe(e, t, n, u) {
  const i = We(e, $(), u == null ? void 0 : u.whiteSpace);
  return Ce(i, t, n);
}
function _t(e, t, n) {
  const u = performance.now(), i = We(e, $(), n == null ? void 0 : n.whiteSpace), r = performance.now(), l = Ce(i, t, !1), c = performance.now();
  let a = 0;
  for (const s of l.breakableWidths)
    s !== null && a++;
  return {
    analysisMs: r - u,
    measureMs: c - r,
    totalMs: c - u,
    analysisSegments: i.len,
    preparedSegments: l.widths.length,
    breakableSegments: a
  };
}
function qt(e, t, n) {
  return Pe(e, t, !1, n);
}
function Zt(e, t, n) {
  return Pe(e, t, !0, n);
}
function Vt(e, t, n) {
  const u = Rt(e, t);
  return { lineCount: u, height: u * n };
}
function ke(e, t, n) {
  let u = n.get(e);
  if (u !== void 0)
    return u;
  u = [];
  const i = be();
  for (const r of i.segment(t[e]))
    u.push(r.segment);
  return n.set(e, u), u;
}
function ye(e) {
  let t = fe.get(e);
  return t !== void 0 || (t = /* @__PURE__ */ new Map(), fe.set(e, t)), t;
}
function Dt(e, t, n, u) {
  return u > 0 && e[u - 1] === "soft-hyphen" && !(t === u && n > 0);
}
function Kt(e, t, n, u, i, r, l) {
  let c = "";
  const a = Dt(t, u, i, r);
  for (let s = u; s < r; s++)
    t[s] === "soft-hyphen" || t[s] === "hard-break" || (s === u && i > 0 ? c += ke(s, e, n).slice(i).join("") : c += e[s]);
  return l > 0 ? (a && (c += "-"), c += ke(r, e, n).slice(u === r ? i : 0, l).join("")) : a && (c += "-"), c;
}
function Me(e, t, n, u, i, r, l) {
  return {
    text: Kt(e.segments, e.kinds, t, u, i, r, l),
    width: n,
    start: {
      segmentIndex: u,
      graphemeIndex: i
    },
    end: {
      segmentIndex: r,
      graphemeIndex: l
    }
  };
}
function Jt(e, t, n) {
  return Me(e, t, n.width, n.startSegmentIndex, n.startGraphemeIndex, n.endSegmentIndex, n.endGraphemeIndex);
}
function Oe(e) {
  return {
    width: e.width,
    start: {
      segmentIndex: e.startSegmentIndex,
      graphemeIndex: e.startGraphemeIndex
    },
    end: {
      segmentIndex: e.endSegmentIndex,
      graphemeIndex: e.endGraphemeIndex
    }
  };
}
function Qt(e, t, n) {
  const u = Gt(e, t, n);
  return u === null ? null : Oe(u);
}
function $t(e, t) {
  return Me(e, ye(e), t.width, t.start.segmentIndex, t.start.graphemeIndex, t.end.segmentIndex, t.end.graphemeIndex);
}
function Xt(e, t, n) {
  return e.widths.length === 0 ? 0 : pe(e, t, (u) => {
    n(Oe(u));
  });
}
function Yt(e, t, n) {
  const u = Qt(e, t, n);
  return u === null ? null : $t(e, u);
}
function en(e, t, n) {
  const u = [];
  if (e.widths.length === 0)
    return { lineCount: 0, height: 0, lines: u };
  const i = ye(e), r = pe(e, t, (l) => {
    u.push(Jt(e, i, l));
  });
  return { lineCount: r, height: r * n, lines: u };
}
function Ut() {
  Qe(), ee = null, fe = /* @__PURE__ */ new WeakMap(), Mt();
}
function tn(e) {
  $e(e), Ut();
}
export {
  Ut as clearCache,
  Vt as layout,
  Yt as layoutNextLine,
  en as layoutWithLines,
  qt as prepare,
  Zt as prepareWithSegments,
  _t as profilePrepare,
  tn as setLocale,
  Xt as walkLineRanges
};
